<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="jennysjaarda" />

<meta name="date" content="2021-10-12" />

<title>Update Meeting: 15-10-2021</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">proxyMR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="setup.html">Setup</a>
</li>
<li>
  <a href="meetings.html">Update meetings</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jennysjaarda/proxyMR">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Update Meeting: 15-10-2021</h1>
<h4 class="author">jennysjaarda</h4>
<h4 class="date">2021-10-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-10-18
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>proxyMR/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210602code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210602)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210602code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210602)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjennysjaardaproxyMRtree28db79b6b958322de072811dc6867bbb41e40a07targetblank28db79ba"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jennysjaarda/proxyMR/tree/28db79b6b958322de072811dc6867bbb41e40a07" target="_blank">28db79b</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjennysjaardaproxyMRtree28db79b6b958322de072811dc6867bbb41e40a07targetblank28db79ba" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jennysjaarda/proxyMR/tree/28db79b6b958322de072811dc6867bbb41e40a07" target="_blank">28db79b</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/meetings.html
    Ignored:    proxyMR_figure_data.RData
    Ignored:    renv/library/
    Ignored:    renv/local/
    Ignored:    renv/staging/

Untracked files:
    Untracked:  R/proxyMR_figures.R
    Untracked:  R/proxyMR_hiercharchy.R
    Untracked:  analysis/project_status.Rmd

Unstaged changes:
    Modified:   analysis/index.Rmd
    Modified:   analysis/todo_list.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/update_meeting_15_10_2021.Rmd</code>) and HTML (<code>docs/update_meeting_15_10_2021.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jennysjaarda/proxyMR/blob/28db79b6b958322de072811dc6867bbb41e40a07/analysis/update_meeting_15_10_2021.Rmd" target="_blank">28db79b</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-10-18
</td>
<td>
fix date in update meeting
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jennysjaarda/proxyMR/28db79b6b958322de072811dc6867bbb41e40a07/docs/update_meeting_15_10_2021.html" target="_blank">28db79b</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-10-18
</td>
<td>
fix date in update meeting
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="sample-definition-issue." class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Sample definition issue.</h1>
<div id="update." class="section level2" number="1.1">
<h2 number="1.1"><span class="header-section-number">1.1</span> Update.</h2>
<p>Noticed that the sample sizes we were working with were quite a bit larger than the Neale database. After a bit of digging, I think I figured out the main reason(s): it seems to be a result of the way we filtered related individuals.</p>
<ul>
<li>First of all, because we’ve done all analyses in each sex separately, I also have performed the related filters in each sex separately. Meaning there is surely related individuals between sexes that remain (of course each individual couple has been filtered separately for their own relatedness).</li>
<li>Second, I am using a function to retain the maximal set of unrelated individuals which is relatively computation heavy and I believe that Neale would have simply filtered based on <code>used_in_pca_calculation</code> from the QC file, which results in &lt;&lt;N.</li>
</ul>
<p>I assume Neale also applied related filtering <em>before</em> PHESANT transformations, while I applied it after (I was thinking that if someone was performing an analysis on a subset of UKB, they would have unnecessarily lost individuals even though the two related individuals weren’t in their subgroup). The problem is, these differences result in slightly different phenotypes in some cases I’ve realized.</p>
<p>For example, phenotype 1050 (amount of time spent outside, hours) is converted by the PHESANT pipeline to an ordinal variable with only 3 categories because the upper categories are so small. Where to split the categories directly depends on the data, with the goal to have roughly equal sized groups. Anyways, in Neale it’s split in different place than in our data.</p>
<div id="discussion-points." class="section level3" number="1.1.1">
<h3 number="1.1.1"><span class="header-section-number">1.1.1</span> Discussion points.</h3>
<ul>
<li>Should we apply the same filter as Neale to get roughly the same sample size and data (particularly for these categorical phenotypes)?
<ul>
<li>Regarding the category definitions: best to have something as similar as possible to Neale’s, because the instrument selection is based on his summary stats, which used his definition. But if it is hard to replicate his definition, it’s not a big problem, it is much more important that we move forward fast</li>
</ul></li>
<li>And/or is it a problem that there we would have related individuals across sexes?
<ul>
<li>For the relateds: unfortunately, they have to be excluded (male-female relatives too, not only same-sex relatives), because when we meta-analyse male-&gt;female and female-&gt;male results, we assume inependence in the estimates, which is not true if you don’t exclude all relatives.</li>
</ul></li>
</ul>
</div>
</div>
<div id="update-2.0." class="section level2" number="1.2">
<h2 number="1.2"><span class="header-section-number">1.2</span> Update 2.0.</h2>
<p>I was able to nearly mimic the sample size of the round 1 GWAS, but not the round 2 GWAS. In the blog post describing the <a href="http://www.nealelab.is/blog/2017/9/11/details-and-considerations-of-the-uk-biobank-gwas">first release</a>, it is mentioned that N = 337,199 individuals are used in the GWAS. I can nearly replicate this number exactly with my data by removing un consenting, non-white and only retaining those that were <code>used_in_pca_calculation</code> (from the genetic QC file). However in the <a href="http://www.nealelab.is/blog/2019/9/16/biomarkers-gwas-results">round 2 blog post</a>, the sample size increases to N = 361,194 individuals.</p>
<p>The details compairing the two samples are found on the <a href="https://github.com/Nealelab/UK_Biobank_GWAS/blob/master/README.md">github page</a>, specifically:</p>
<ul>
<li><strong>imputed-v3 parameters</strong>
<ul>
<li>Used.in.pca.calculation filter (unrelated samples)</li>
<li>sex chromosome aneuploidy filter</li>
<li>Use provided PCs for European sample selection to determine British ancestry
<ul>
<li>Use 7 standard deviations away from the 1st 6 PCs</li>
<li>Further Filter to self-reported ‘white-British’ / ‘Irish’ / ‘White’</li>
</ul></li>
<li><strong>QCed sample count: 361194 samples</strong></li>
</ul></li>
<li><strong>imputed-v2 parameters</strong>
<ul>
<li>Used.in.pca.calculation filter (unrelated samples)</li>
<li>sex chromosome aneuploidy filter</li>
<li>White.british.ancestry filter</li>
<li><strong>QCed sample count: 337199 samples</strong></li>
</ul></li>
</ul>
<div id="discussion-points.-1" class="section level3" number="1.2.1">
<h3 number="1.2.1"><span class="header-section-number">1.2.1</span> Discussion points.</h3>
<p>Our current filtering matches closely the v2 filtering. If we tried to mimic the v3 filtering, we could increase our sample size. What do we think about this discrepency? And more importantly, the decrease in power due to smaller sample size (about 25K less than the v3 parameters)?</p>
</div>
</div>
</div>
<div id="progress-update." class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Progress update.</h1>
<div id="phenotype-selection." class="section level2" number="2.1">
<h2 number="2.1"><span class="header-section-number">2.1</span> Phenotype selection.</h2>
<ul>
<li>Added back binary phenotypes.</li>
<li>Removed all dietary phenotypes (binary, ordinal and continuous).</li>
<li>Removed some redundant phenotypes:
<ul>
<li>Left-side body traits (highly correlated with right-side) were removed</li>
<li>All ‘qualifications’ data (corresponding to field <a href="https://biobank.ctsu.ox.ac.uk/showcase/field.cgi?id=6138">6138</a>) because data is roughly captured with continuous variables (such as years of education).</li>
</ul></li>
<li>See <a href="phenotype_selection.html">phenotype selection</a> document for complete summary of selection and remaining phenotypes.</li>
</ul>
</div>
<div id="two-trait-mr-filtering." class="section level2" number="2.2">
<h2 number="2.2"><span class="header-section-number">2.2</span> Two-trait MR filtering.</h2>
<p>Restrict to only XY pairs that are have correlation &lt; <code>0.8</code>. Correlations were calculated in the full biobank (unrelated, British, consenting etc.), not just within couples, in both sexes combined. Trait pairs and their correlations are summarized in the table below.</p>
<p>In total, <code>#r length(which(household_MR_summary.DT$corr_traits &gt; 0.8))</code> trait pairs were removed due to correlation &gt; <code>0.8</code>.</p>
<p><strong>Note</strong>: after filtering out pairs with high correlation, a bonferroni correction applied (p &lt; 0.05 / [num remaining pairs after correaltion filter]) before examining paths from <span class="math inline">\(X_i \rightarrow Y_p\)</span> (i.e we only examined pairs that were significant along the diagonal).</p>
<p><img src="assets/two_trait_model.png" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-two_trait_model">
Past versions of two_trait_model.png
</button>
</p>
<div id="fig-two_trait_model" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jennysjaarda/proxyMR/blob/fda484320b96a92176e7aa7e0d8c3d64000bd513/docs/assets/two_trait_model.png" target="_blank">fda4843</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-09-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="adjust-y_i-rightarrow-y_p-for-effect-of-x_i-rightarrow-y_p-in-mvmr." class="section level2" number="2.3">
<h2 number="2.3"><span class="header-section-number">2.3</span> Adjust <span class="math inline">\(Y_i \rightarrow Y_p\)</span> for effect of <span class="math inline">\(X_i \rightarrow Y_p\)</span> in MVMR.</h2>
<p>Previously, I had adjusted the <span class="math inline">\(Y_i \rightarrow Y_p\)</span> MR results for effects of <span class="math inline">\(X_i\)</span> (as depicted below), but had only used <span class="math inline">\(Y\)</span> IVs.</p>
<p><img src="assets/two_trait_model_yiyp_adj.png" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-two_trait_model_yiyp_adj">
Past versions of two_trait_model_yiyp_adj.png
</button>
</p>
<div id="fig-two_trait_model_yiyp_adj" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jennysjaarda/proxyMR/blob/529020ff80300880fe93b93913dba7934bc26400/docs/assets/two_trait_model_yiyp_adj.png" target="_blank">529020f</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-09-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Decided best to also include all <span class="math inline">\(X\)</span> IVs in the model as well. MRMR model is: <span class="math inline">\(Y_p \sim Y_i + X_i\)</span>.</p>
<p>To ensure a set of independent SNPs, <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> IVs were clumped using at a threshold of <span class="math inline">\(r^2\)</span> &lt; 0.001 using the 1000 Genomes European data set. Clumping p-values for priorization of SNPs, were based on the Neale summary statistics with <span class="math inline">\(Y_i\)</span> in both sexes, ensuring the same set of SNPs were selected in the male and female models.</p>
</div>
<div id="mvmr" class="section level2" number="2.4">
<h2 number="2.4"><span class="header-section-number">2.4</span> MVMR</h2>
<p>For each <span class="math inline">\(X_i \rightarrow Y_p\)</span> MR, we included relevant covariates (Z’s) in a MVMR framework in the <span class="math inline">\(X_p \rightarrow Y_p\)</span> MR. Z’s were selected based on having BF significance in <strong>both</strong> of the MRs:</p>
<ol style="list-style-type: decimal">
<li><span class="math inline">\(X \rightarrow Z\)</span></li>
<li><span class="math inline">\(Z \rightarrow Y\)</span></li>
</ol>
<p>After Z’s were selected, they were further pruned based on phenotypic correlation (same as above). First, any Z with phenotypic correlation with <span class="math inline">\(X\)</span> or <span class="math inline">\(Y\)</span> &gt; 0.8 was removed. Second, beginning with the most significant <span class="math inline">\(Z_j\)</span> association with <span class="math inline">\(Y\)</span> (i.e. <span class="math inline">\(Z \rightarrow Y\)</span>), all reamining Z’s with phenoytpic correlation &gt; 0.8 were removed. This process was repeated until all remaining Z’s had phenotypic correlation &lt; 0.8 with each other and <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>.</p>
<p>The resulting MVMR model is: <span class="math inline">\(Y \sim X + Z_1 + ... + Z_k\)</span> run in same individual (see figure).</p>
<p><img src="assets/MV_trait_model.png" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-MV_trait_model">
Past versions of MV_trait_model.png
</button>
</p>
<div id="fig-MV_trait_model" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jennysjaarda/proxyMR/blob/1afb033cbb3ccd74817ca47d915a120c41c00432/docs/assets/MV_trait_model.png" target="_blank">1afb033</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-09-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Finally, IVs for all <span class="math inline">\(Z\)</span> and <span class="math inline">\(X\)</span> were clumped as above (using at a threshold of <span class="math inline">\(r^2\)</span> &lt; 0.001 using the 1000 Genomes European data set. Clumping p-values for priorization of SNPs, were based on the Neale summary statistics with <span class="math inline">\(X_i\)</span> in both sexes, ensuring the same set of SNPs were selected in the male and female models).</p>
</div>
<div id="pc-impact" class="section level2" number="2.5">
<h2 number="2.5"><span class="header-section-number">2.5</span> PC impact</h2>
<p>Tested the following: - cor(PC_i,PC_p) for each PC (i.e. within couple PC correlation). - cor(X,PC) for all PCs.</p>
<p>PC correlations among couples is shown below.</p>
<p>Correlation of PCs with traits in pipeline is shown below.</p>
<p>Next, I calculated the correlation due to counfounding as <code>cor(X,PC)^2*cor(PC_i,PC_p)</code> values, and plotted them against the raw <code>cor(X_i,X_p)</code> values.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.0 (2021-05-18)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Big Sur 11.5.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] kableExtra_1.3.4 knitr_1.33       DT_0.18          forcats_0.5.1   
 [5] stringr_1.4.0    dplyr_1.0.7      purrr_0.3.4      readr_2.0.1     
 [9] tidyr_1.1.3      tibble_3.1.2     ggplot2_3.3.5    tidyverse_1.3.1 
[13] targets_0.7.0    workflowr_1.6.2 

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7        svglite_2.0.0     lubridate_1.7.10  ps_1.6.0         
 [5] assertthat_0.2.1  rprojroot_2.0.2   digest_0.6.27     utf8_1.2.1       
 [9] R6_2.5.0          cellranger_1.1.0  backports_1.2.1   reprex_2.0.0     
[13] evaluate_0.14     highr_0.9         httr_1.4.2        pillar_1.6.1     
[17] rlang_0.4.11      readxl_1.3.1      rstudioapi_0.13   data.table_1.14.0
[21] whisker_0.4       callr_3.7.0       rmarkdown_2.10    webshot_0.5.2    
[25] htmlwidgets_1.5.3 igraph_1.2.6      munsell_0.5.0     broom_0.7.6      
[29] compiler_4.1.0    httpuv_1.6.1      modelr_0.1.8      xfun_0.23        
[33] systemfonts_1.0.2 pkgconfig_2.0.3   htmltools_0.5.1.1 tidyselect_1.1.1 
[37] codetools_0.2-18  viridisLite_0.4.0 fansi_0.5.0       crayon_1.4.1     
[41] tzdb_0.1.2        dbplyr_2.1.1      withr_2.4.2       later_1.2.0      
[45] grid_4.1.0        jsonlite_1.7.2    gtable_0.3.0      lifecycle_1.0.0  
[49] DBI_1.1.1         git2r_0.28.0      magrittr_2.0.1    scales_1.1.1     
[53] cli_2.5.0         stringi_1.6.2     renv_0.13.2-62    fs_1.5.0         
[57] promises_1.2.0.1  xml2_1.3.2        ellipsis_0.3.2    generics_0.1.0   
[61] vctrs_0.3.8       tools_4.1.0       glue_1.4.2        hms_1.1.0        
[65] processx_3.5.2    yaml_2.2.1        colorspace_2.0-1  rvest_1.0.0      
[69] haven_2.4.1      </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
