<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="jennysjaarda" />

<meta name="date" content="2021-06-14" />

<title>Update Meeting: 14-06-2021</title>

<script src="site_libs/header-attrs-2.8/header-attrs.js"></script>
<script src="site_libs/jquery-3.5.1/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.18/datatables.js"></script>
<link href="site_libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">proxyMR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="setup.html">Setup</a>
</li>
<li>
  <a href="meeting.html">Update meetings</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jennysjaarda/proxyMR">
    <span class="fas fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Update Meeting: 14-06-2021</h1>
<h4 class="author">jennysjaarda</h4>
<h4 class="date">2021-06-14</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-07-22
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>proxyMR/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210602code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210602)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210602code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210602)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjennysjaardaproxyMRtree38566523b07bbf35e343879f601ba3c2df7f9477targetblank3856652a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jennysjaarda/proxyMR/tree/38566523b07bbf35e343879f601ba3c2df7f9477" target="_blank">3856652</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjennysjaardaproxyMRtree38566523b07bbf35e343879f601ba3c2df7f9477targetblank3856652a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jennysjaarda/proxyMR/tree/38566523b07bbf35e343879f601ba3c2df7f9477" target="_blank">3856652</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    renv/library/
    Ignored:    renv/local/
    Ignored:    renv/staging/

Untracked files:
    Untracked:  analysis/latex_test.Rmd
    Untracked:  household_MR_exhaustive_summary.rds
    Untracked:  summary.RData
    Untracked:  traits_corr2_update.rds

Unstaged changes:
    Modified:   R/functions.R
    Modified:   _targets.R
    Modified:   analysis/update_meeting_05_07_2021.Rmd
    Modified:   analysis/update_meeting_14_07_2021.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/update_meeting_14_06_2021.Rmd</code>) and HTML (<code>docs/update_meeting_14_06_2021.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jennysjaarda/proxyMR/blob/38566523b07bbf35e343879f601ba3c2df7f9477/analysis/update_meeting_14_06_2021.Rmd" target="_blank">3856652</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-07-22
</td>
<td>
wflow_publish(“analysis/update_meeting_14_06_2021.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jennysjaarda/proxyMR/021aeb825841b4875c7a17a2c7539bfc531fc5dc/docs/update_meeting_14_06_2021.html" target="_blank">021aeb8</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-06-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jennysjaarda/proxyMR/blob/14a8a7a35de376198751e25986f562b742542a81/analysis/update_meeting_14_06_2021.Rmd" target="_blank">14a8a7a</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-06-18
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/update_meeting_14_06_2021.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jennysjaarda/proxyMR/b02d111a9fe4eb009dcf1aa3e27d7e839414788c/docs/update_meeting_14_06_2021.html" target="_blank">b02d111</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-06-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jennysjaarda/proxyMR/blob/dfe878d42218ff0e2fa49f70b15262679f69cb39/analysis/update_meeting_14_06_2021.Rmd" target="_blank">dfe878d</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-06-18
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/update_meeting_14_06_2021.Rmd”))
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jennysjaarda/proxyMR/blob/750d442ecb54dc5268e4c2557e59cae052011d0a/analysis/update_meeting_14_06_2021.Rmd" target="_blank">750d442</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-06-16
</td>
<td>
add website docs
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jennysjaarda/proxyMR/750d442ecb54dc5268e4c2557e59cae052011d0a/docs/update_meeting_14_06_2021.html" target="_blank">750d442</a>
</td>
<td>
jennysjaarda
</td>
<td>
2021-06-16
</td>
<td>
add website docs
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="last-week-summary." class="section level2">
<h2>Last week summary.</h2>
<div id="next-steps" class="section level3">
<h3>Next steps:</h3>
<ol style="list-style-type: decimal">
<li><p>Summarize all our results to get a feel for top traits (let’s call this: trait/trait MR). A logical addition that we didn’t discuss would be to meta-analyze across sexes and see what traits are significant in pooled analysis and then check which of these are significantly different between sexes.</p></li>
<li><p>Test linear association across bins (both age and time-together bins) using IVW estimated inversely weighted by SE.</p></li>
<li><p>Combine age and time-together bins into a grid of effect sizes to see if we see any pattern.</p></li>
<li><p>Move beyond our trait/trait MR into investigating how one trait impacts other traits to tease out direct vs indirect effects, two parts here:</p>
<ol style="list-style-type: lower-alpha">
<li>as univariate model and</li>
<li>as multivariate model (selecting traits for multivariate model that were nominal in univariate).</li>
</ol></li>
<li><p>Replicate permutation procedure from Tenesa paper. Do you still want to explore this?</p></li>
<li><p>Explore dietary traits - request estimates from Ninon?</p></li>
</ol>
</div>
<div id="follow-up-questions" class="section level3">
<h3>Follow-up questions:</h3>
<p>Re (#4), for the trait/trait MR, I essentially followed a 2-sample MR procedure (just as a refresher for us both), so for example BMI partner vs BMI trait:</p>
<ul>
<li>Find IVs for BMI in Neale and take these estimates as betaX.</li>
<li>Test association of IVs with partner BMI in couples (this is betaY).</li>
<li>Combine effects together in standard MR framework (betaY ~ betaX).</li>
</ul>
<p>For the univariate analysis between two traits, I think what you were suggesting is to follow exactly the same procedure as above. So for example, disease partner vs BMI trait:</p>
<ul>
<li>Same IVs as above and take these estimates as betaX.</li>
<li>Test association of these IVs with partner “different phenotype” (some disease, let’s say) (betaY).</li>
<li>Combine in MR framework.</li>
</ul>
<p>For the multivariate model, so if the model of interest is BMI ~ BMI + education + diet + activity level:</p>
<ul>
<li>IVs are the union of the IVs for each individual trait (estimates taken from Neale: beta_BMI, beta_edu, beta_diet).</li>
<li><strong>Additional step</strong>: prune these SNPs to ensure independence.</li>
<li>Test association of these pruned SNPs with BMI in couples (betaY).</li>
<li>Combine effects in multivariate MR framework: betaY ~ beta_BMI + beta_edu + beta_diet, etc…</li>
</ul>
<p>Does this look right (particularly for the multivariate model)? i.e. the final MR model only includes beta estimates, no raw genetic data.</p>
<p>I also understood that one of the traits to explore would be PCs. To me it seems logical that this would be a causal variable to explore (for eg, an MR being: BMI_beta ~ PCX_beta) rather than the other way around. If I have that right, then that means we would need IVs for PCs. At quick glance, Neale doesn’t have this available. Do you know if anyone in the group would have done this, otherwise I should run this GWAS.</p>
</div>
<div id="responses" class="section level3">
<h3>Responses:</h3>
<ul>
<li>Re. #3: also test heterogeneity (in case there is discrepancy, but no trend), using Cochran’s Q stat.</li>
<li>Re. combining time-together and age bins: Do this only if the previous point gave something promising. Otherwise drop it.</li>
<li>Not urgent to replicate Tenesa paper.</li>
<li>To explore dietary traits, got in touch again with Nicola Pirastu.</li>
<li>Raw genetic data is only used to derive the betaY, but from then it is a 2-sample MR.</li>
<li>PCs are only exposures, not outcomes. Need to run GWAS.</li>
</ul>
</div>
</div>
<div id="results." class="section level2">
<h2>Results.</h2>
<div id="traits-analyzed." class="section level3">
<h3>Traits analyzed.</h3>
</div>
<div id="summary-of-results" class="section level3">
<h3>Summary of results</h3>
<ul>
<li>IVW estimates report sex-specific results</li>
<li>MR_meta statistics refer to the meta of both-sexes combined.</li>
<li>Sex heterogeneity column reports the p-value of the difference between the MR estimates from the two sexes.</li>
</ul>
</div>
<div id="testing-discrepancy-amongst-bins." class="section level3">
<h3>Testing discrepancy amongst bins.</h3>
<ul>
<li>The first table represents the results of individuals binned by average age.</li>
<li>The second table represent the results of individuals binned by time together.</li>
<li>Q p-values report the difference amongst bins.</li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.0 (2021-05-18)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Big Sur 11.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] knitr_1.33         DT_0.18            forcats_0.5.1      stringr_1.4.0     
 [5] dplyr_1.0.6        purrr_0.3.4        readr_1.4.0        tidyr_1.1.3       
 [9] tibble_3.1.2       ggplot2_3.3.3      tidyverse_1.3.1    targets_0.5.0.9000
[13] workflowr_1.6.2   

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6        lubridate_1.7.10  ps_1.6.0          assertthat_0.2.1 
 [5] rprojroot_2.0.2   digest_0.6.27     utf8_1.2.1        R6_2.5.0         
 [9] cellranger_1.1.0  backports_1.2.1   reprex_2.0.0      evaluate_0.14    
[13] httr_1.4.2        pillar_1.6.1      rlang_0.4.11      readxl_1.3.1     
[17] rstudioapi_0.13   data.table_1.14.0 whisker_0.4       callr_3.7.0      
[21] rmarkdown_2.8     htmlwidgets_1.5.3 igraph_1.2.6      munsell_0.5.0    
[25] broom_0.7.6       compiler_4.1.0    httpuv_1.6.1      modelr_0.1.8     
[29] xfun_0.23         pkgconfig_2.0.3   htmltools_0.5.1.1 tidyselect_1.1.1 
[33] codetools_0.2-18  fansi_0.5.0       crayon_1.4.1      dbplyr_2.1.1     
[37] withr_2.4.2       later_1.2.0       grid_4.1.0        jsonlite_1.7.2   
[41] gtable_0.3.0      lifecycle_1.0.0   DBI_1.1.1         git2r_0.28.0     
[45] magrittr_2.0.1    scales_1.1.1      cli_2.5.0         stringi_1.6.2    
[49] renv_0.13.2-62    fs_1.5.0          promises_1.2.0.1  xml2_1.3.2       
[53] ellipsis_0.3.2    generics_0.1.0    vctrs_0.3.8       tools_4.1.0      
[57] glue_1.4.2        crosstalk_1.1.1   hms_1.1.0         processx_3.5.2   
[61] yaml_2.2.1        colorspace_2.0-1  rvest_1.0.0       haven_2.4.1      </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
