---
title: "Manuscript"
author: "Jenny Sjaarda"
date: "2022-01-04"
output: workflowr::wflow_html
---

```{r setup, include = FALSE}

options(scipen=999)
source("code/settings.R")
source("R/functions.R")
library(targets)
library(kableExtra)
library(DT)
library(knitr)
library(ggplot2)
library(kableExtra)
library(cowplot)
library(plotly)
library(plyr)
library(dplyr)
library(ggrepel)
library(viridis)
library(hrbrthemes)
library(ggpubr)
library(broom)
library(data.table)
library(xlsx)
```

```{r load_targets, include = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE)}

tar_load(Neale_SGG_dir_filt)
tar_load(traits_corr_spearman)
tar_load(couple_MR_vs_trait_corr)
tar_load(corr_impact_by_PCs)
tar_load(corr_impact_by_traits)
tar_load(corr_impact_by_coords)
tar_load(corr_potential_trait_confounders_pos)
tar_load(corr_potential_trait_confounders_pos_MVMR_summary)
tar_load(traits_corr)
tar_load(PCs_corr)
tar_load(traits_corr2_filled)
tar_load(household_MR_summary_AM)
tar_load(num_tests_by_PCs)
tar_load(num_tests_by_PCs_AM_sig)
tar_load(household_MR_binned_het_sex_spec)
tar_load(household_MR_binned_het_SNPmeta)
tar_load(traits_final)
tar_load(household_MR_AM_FvsM_fig)
tar_load(confounder_traits)
tar_load(proxyMR_prod_comparison_fig_yiyp_adj)
tar_load(proxyMR_comparison_summary_yiyp_adj_SNPmeta)
tar_load(household_MR_summary_AM)
tar_load(household_MR_binned_het_SNPmeta)
#tar_load(potential_trait_confounders_MVMR)
tar_load(corr_mat_traits)
tar_load(corr_impact_by_traits_causal)
tar_load(household_MR_SNPmeta_sensitivity)
tar_load(binned_pheno_corrs)


```

# A few notes (for myself). 

- The number of effective tests can be found at the target `num_tests_by_PCs`. 
- The number of effective tests among those significant in the MR can be found at the target `num_tests_by_PCs_AM_sig`.

# Pearson vs. Spearman correlation.

We wanted to ensure that INQT was not significantly impacting the correaltions of each trait, so we also calculated a non-parametric based correaltion (Spearman) and found consistent correlations.  

```{r pearson_vs_spearman, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), message=FALSE, warning = FALSE}

Neale_SGG_dir_filt_unique <- Neale_SGG_dir_filt[,c("phenotype", "phenotype_sub", "sgg_phesant_name", "description", "SGG_available", "variable_type")] %>% unique()

## looks at all traits *before* the correlation filter

pearson_vs_spearman_dat <-
  inner_join(traits_corr,
             unique(Neale_SGG_dir_filt_unique[, c("sgg_phesant_name", "variable_type")]),
             by = c("ID" = "sgg_phesant_name")) %>%
  dplyr::rename(r2_pearson = r2) %>%
  inner_join(traits_corr_spearman[, c("ID", "r2")]) %>% dplyr::rename(r2_spearman = r2) %>%
  as_tibble() %>% type_convert() %>%
  mutate(r_pearson = sqrt(r2_pearson)) %>%
  mutate(r_spearman = sqrt(r2_spearman)) 

pearson_vs_spearman_plot <-
  ggplot2::ggplot(data = pearson_vs_spearman_dat,
                  ggplot2::aes(
                    x = r_pearson,
                    y = r_spearman,
                    color = factor(variable_type)
                  )) +
  geom_point() +
  theme_minimal() +
  ggplot2::labs(x = paste0("Pearson correlation"),
                y = paste("Spearman correlation")) +
  scale_color_manual(
    name = "Variable type",
    values = custom_col[c(1, 2, 3)],
    labels = c("Binary", "Continuous (IRNT)", "Ordinal")
  )

pearson_vs_spearman_plot
```

# Relationship between causal effects and raw phenotypic correlation in couples.

We sought to compare the raw correlation amongst couples vs. the standardized MR effects. There are many traits where the `correlation > MR effects` including some traits of interest such as standing height, place of birth, among many others. 

For example, if we take a trait such as place of birth (NC), this is very highly correlated among couples. However, the within-couple MR-estimate is smaller than the observed phenotypic correaltion. This is the case for many such traits. 

```{r pheno_corr_plot, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), message=FALSE, warning = FALSE}

### phenotypic correlation plot for 129_irnt (place of birth) 

tar_load(path_pheno_data)
tar_load(hh_pairs_filter)
Neale_pheno_ID <- "129_irnt" ## this is the Neale_id, used to be pheno_description
phes_ID <- gsub("_irnt", "", Neale_pheno_ID)

male_pheno_file <- path_pheno_data[which(endsWith(path_pheno_data, paste0("/phesant/", Neale_pheno_ID, "_male.txt")))]
female_pheno_file <- path_pheno_data[which(endsWith(path_pheno_data, paste0("/phesant/", Neale_pheno_ID, "_female.txt")))]
male_pheno_data <- fread(male_pheno_file, header=TRUE, select=c("IID","sex","age", phes_ID))
female_pheno_data <- fread(female_pheno_file, header=TRUE, select=c("IID","sex","age", phes_ID))

joint_data <- rbind(male_pheno_data, female_pheno_data)


pairs_filter_copy <- hh_pairs_filter
    pairs_filter_copy <- pairs_filter_copy %>%
      mutate(HOUSEHOLD_MEMBER1_M = case_when(HOUSEHOLD_MEMBER1_sex == 1 ~ HOUSEHOLD_MEMBER1,
                                             HOUSEHOLD_MEMBER2_sex == 1 ~ HOUSEHOLD_MEMBER2)) %>%

      mutate(HOUSEHOLD_MEMBER2_F = case_when(HOUSEHOLD_MEMBER1_sex == 0 ~ HOUSEHOLD_MEMBER1,
                                             HOUSEHOLD_MEMBER2_sex == 0 ~ HOUSEHOLD_MEMBER2)) %>%

      dplyr::select(-HOUSEHOLD_MEMBER1, -HOUSEHOLD_MEMBER2) %>%
      dplyr::rename(HOUSEHOLD_MEMBER1 = HOUSEHOLD_MEMBER1_M) %>%
      dplyr::rename(HOUSEHOLD_MEMBER2 = HOUSEHOLD_MEMBER2_F)

  
pairs_filter_copy$trait1 <- joint_data[[phes_ID]][match(pairs_filter_copy[["HOUSEHOLD_MEMBER1"]], joint_data$IID)]
pairs_filter_copy$trait2 <- joint_data[[phes_ID]][match(pairs_filter_copy[["HOUSEHOLD_MEMBER2"]], joint_data$IID)]

pheno_corr_dat <- pairs_filter_copy

pheno_corr_plot <- ggplot2::ggplot(data = pheno_corr_dat, ggplot2::aes(x = trait1,
                                                       y = trait2)) +
  geom_point(alpha = 1.2, color = "gray") +
  #geom_text(show.legend = FALSE) +
  #geom_abline(slope=1, intercept=0) +
  
  theme_minimal() +

  #geom_text(aes(x=1,y=2,label=""),show.legend=F) +
  ggplot2::labs(x = paste0("Male place of birth (NC)"),
                y = paste("Female place of birth (NC)")) + 
  geom_smooth(method=lm , color="red", fill="#69b3a2", se=TRUE) +
  theme_ipsum()

pheno_corr_plot

```

## Summary of cases where correlation is different from the MR estimate. 

A plot comparing the two effects, and a summary table of the significant effects is shown below. 

```{r corr_vs_MR_effects, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), message=FALSE}

interesting_phenotypes <- c("Place of birth in UK - north co-ordinate", 
                            "Systolic blood pressure, automated reading", "Standing height", 
                            "Weight", "Forced vital capacity (FVC), Best measure",
                            "Basophill count", "Pulse rate, automated reading")


corr_vs_MR_effects_dat <- replace_Neale_ID(couple_MR_vs_trait_corr, traits_corr2_filled, "exposure_ID", "outcome_description") %>% 
  mutate(Phenotype = outcome_description) %>% 
  mutate(Phenotype = case_when(Phenotype == "Basal metabolic rate" ~ "BMR",
                               Phenotype == "Systolic blood pressure, automated reading" ~ "SBP",
                               Phenotype == "Pulse rate, automated reading" ~ "Pulse rate",
                               Phenotype == "Body fat percentage" ~ "BFP",
                               Phenotype == "Standing height" ~ "Height",
                               Phenotype == "Reticulocyte percentage" ~ "Reticulocyte %",
      
                               #Phenotype == "Weight" & outcome_ID == "21002_irnt" ~ "Weight",
                               Phenotype == "Forced vital capacity (FVC), Best measure" ~ "FVC",
                               Phenotype == "Red blood cell (erythrocyte) count" ~ "RBC count",
                               Phenotype == "Place of birth in UK - north co-ordinate" ~ "Place of birth, NC",
                               TRUE ~ Phenotype)) %>%
  
  dplyr::rename(Correlation = couple_r) %>% dplyr::rename(MR_est = IVW_beta) %>% 
  mutate(diff_p_sig = case_when(diff_p < 0.05/num_tests_by_PCs ~ TRUE, TRUE ~ FALSE)) %>%
  mutate(outcome_description2 = case_when(outcome_description=="Weight" & outcome_ID == "23098_irnt" ~ "Weight2",
                                          TRUE ~ outcome_description)) %>%
  mutate(label_phenos = case_when(outcome_description2 %in% interesting_phenotypes ~ TRUE, TRUE ~ FALSE))


  # TRUE - BF sig difference 
  # FALSE - non-sig 
  #mutate(diff_p_sig = ifelse(diff_p < 0.05/num_tests_by_PCs, TRUE, FALSE))

corr_vs_MR_effects_plot <- ggplot2::ggplot(data = corr_vs_MR_effects_dat, ggplot2::aes(x = Correlation,
                                                       y = MR_est, 
                                                       label = Phenotype, 
                                                       color = factor(diff_p_sig)),) +
  
  ggplot2::geom_errorbar(
    ggplot2::aes(ymin = MR_est - 1.96*IVW_se, ymax = MR_est + 1.96*IVW_se),
    colour = "grey",
    width = 0
  ) +
  ggplot2::geom_errorbar(
    ggplot2::aes(xmin = Correlation - 1.96*couple_r_se, xmax = Correlation + 1.96*couple_r_se),
    colour = "grey",
    width = 0
  ) +
  
  geom_point(alpha = 1.2) +
  scale_color_manual(values = custom_col[c(1,2)], labels=c("Non-significant","Bonferroni-significant difference")) +
  #geom_text(show.legend = FALSE) +
  geom_abline(slope=1, intercept=0) +

  
  theme_minimal() +
  theme(legend.text=element_text(size=11)) +
  #geom_text(aes(x=1,y=2,label=""),show.legend=F) +
  ggplot2::labs(x = paste0("Couple correlation"),
                y = paste("MR-estimate")) +
  theme(legend.position="bottom") + theme(legend.title=element_blank()) +
  geom_text_repel(data = subset(corr_vs_MR_effects_dat, label_phenos==TRUE),
                  nudge_y = -0.35 - subset(corr_vs_MR_effects_dat, label_phenos==TRUE)$MR_est, size = 3.25,
                  direction = "x", show.legend = FALSE, box.padding = 0.5 ) 

corr_vs_MR_effects_plot


corr_vs_MR_effects_sig_dat <- corr_vs_MR_effects_dat %>% dplyr::rename(IVW_beta = MR_est) %>% 
  filter(diff_p < 0.05/num_tests_by_PCs) %>%
  mutate(mr_corr_diff = Correlation - IVW_beta) %>% 
  mutate(correlation_larger = ifelse(Correlation > IVW_beta, TRUE, FALSE)) %>% 
  arrange(diff_p) %>% dplyr::select(-exposure_ID, -Phenotype, -outcome_description2, -label_phenos, -mr_corr_diff, -diff_p_sig) %>% 
  dplyr::rename(trait_description = outcome_description) %>%
  dplyr::rename(trait_ID = outcome_ID) %>%
  dplyr::rename(couple_r = Correlation) %>%
  dplyr::select(trait_description, trait_ID, everything())

## prune for correlations < 0.8

corr_vs_MR_effects_sig_dat.DT <- corr_vs_MR_effects_sig_dat %>% 
  dplyr::select(trait_description, trait_ID, couple_r, IVW_beta, IVW_pval, diff_p, correlation_larger)

#prune_pheno_table(corr_vs_MR_effects_sig_dat, "trait_ID", corr_mat_traits, corr_trait_threshold)

# Could prune the results, but only one trait is removed so seems pointless: 

# Result are pruned to only show traits with correlation < `r corr_trait_threshold`. (The number of traits before pruning was `r dim(corr_vs_MR_effects_sig_dat)[1]` vs `r dim(corr_vs_MR_effects_sig_dat_pruned.DT)[1]` after pruning).
```

The table below displays the `r dim(corr_vs_MR_effects_sig_dat.DT)[1]` traits that have correlation significantly different from the MR-estimate. Of the significant difference `r table(corr_vs_MR_effects_sig_dat$correlation_larger)[[1]]` have MR-estimates larger than than correlation, suggesting the presence of a negative confounder (this table corresponds to "sig_corr_vs_MR" in the excel document).

```{r corr_vs_MR_effects_table, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), message=FALSE}

format_round_cols <- colnames(dplyr::select_if(corr_vs_MR_effects_sig_dat.DT, is.numeric))

DT::datatable(corr_vs_MR_effects_sig_dat.DT %>% dplyr::select(-trait_ID), width = "100%") %>% formatSignif(columns=format_round_cols, digits = 3)

```

## Search for potential confounders to explain cases where correlation differs from MR estimate. 

Sought to identify potential confounders (from our list of traits in the pipeline) which may, in part, explain the discrepant estimates, as follows:

1. Restrict to only traits that have correlation significantly different from MR-estimate (`r dim(corr_vs_MR_effects_sig_dat)[1]` traits have correlation different from MR_estimate).
1. Remove any `potential_counders` with correlation with `trait_ID` > `r corr_trait_threshold`.
1. For each of the remaining $Z$ `potential_counders` for each trait $X$, find the potential confounders which have a causal effect on trait $X$ ($\alpha_{z\rightarrow x}$ p-val < 0.05/[number of effective tests]).
1. Split results into those cases where MR-estimate is **greater** than the phenotypic correlation and those that are **less** than the correlation. 
1. For traits where correlation was less than MR-estimate, we restricted to negative confounders (i.e. negative $\alpha_{z_i\rightarrow z_p}$), and conversely for traits where the correlation was greater than the MR-estimate we restricted to positive confounders (i.e. positive $\alpha_{z_i\rightarrow z_p}$) [note that only one significant MR-effect is negative].  
1. Of the remaining $Z$, we further filtered to those with a significant within couple MR effect ($\alpha_{z_i\rightarrow z_p}$ p < 0.05/[number of remaining $Z$]). 
1. Calculate the correlation due to confounding as: $C = (\alpha_{z\rightarrow x})^2 * \alpha_{z_i\rightarrow z_p}$.
1. Calculate the ratio of this correlation ($C$) and the correlation of $X$ in partners ($C/r_x$). 
1. For each `trait_ID` prune remaining `potential_counders` so that all have correlations <  `r corr_trait_threshold`, prioritized by their `corr_due_to_confounding_ratio`. 

```{r simple_C_model, echo = FALSE}
library(knitr)
knitr::include_graphics("assets/simple_confounder_model.png", error = FALSE)
```

A summary of the top 5 confounders for each `trait_ID` can be seen below (a variation of this table, include *all* potential confounders is included in the excel sheet titled "sig_confounders"): 

```{r confounder_summary, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE}

number_of_potential_confounders <- dim(corr_potential_trait_confounders_pos)[1]/dim(corr_potential_trait_confounders_pos)[1]

non_sig_confounders <- corr_potential_trait_confounders_pos %>% filter(!sig_confounder) %>% group_by(outcome_ID) %>%
  dplyr::summarize(description = outcome_description[1], n = n()) %>% 
  filter(n==number_of_potential_confounders) %>%
  mutate(n = n-number_of_potential_confounders) %>% mutate(potential_confounders = NA) %>%
  mutate(mean_C_ratio = NA) %>%
  dplyr::rename(trait_ID = outcome_ID) 

sig_confounders <- corr_potential_trait_confounders_pos_MVMR_summary %>% filter(sig_confounder) %>% group_by(outcome_ID) %>%
  arrange(-corr_due_to_confounding_MVMR) %>%
  mutate(total_potential_confounders = n()) %>%
  dplyr::summarize(description = outcome_description[1],
                   total_potential_confounders = total_potential_confounders[1],
                   potential_confounders = paste0(exposure_description, collapse = "; "), 
                   max_C = max(corr_due_to_confounding_MVMR)) %>%
  dplyr::rename(trait_ID = outcome_ID) %>%
  dplyr::rename(trait_description = description) %>% 
  dplyr::select(trait_description, trait_ID, everything())

sig_confounders_top5 <- corr_potential_trait_confounders_pos_MVMR_summary %>% filter(sig_confounder) %>% group_by(outcome_ID) %>%
  arrange(-corr_due_to_confounding_MVMR) %>%
  mutate(total_potential_confounders = n()) %>%
  slice_head(n = 5) %>%
  #mutate(potential_confounders = paste0(exposure_description, collapse = "; ")) %>%
  dplyr::summarize(trait_description = outcome_description[1],
                   total_potential_confounders = total_potential_confounders[1],
                   top5_potential_confounders = paste0(exposure_description, collapse = "; "), 
                   max_C = max(corr_due_to_confounding_MVMR)) %>%
  dplyr::rename(trait_ID = outcome_ID) %>%
  left_join(corr_vs_MR_effects_sig_dat.DT, by = c("trait_description", "trait_ID")) %>%
  dplyr::select(trait_description, trait_ID, couple_r, IVW_beta, IVW_pval,everything())

format_round_cols <- colnames(dplyr::select_if(sig_confounders_top5, is.numeric))

DT::datatable(sig_confounders_top5, width = "100%") %>% formatSignif(columns=format_round_cols, digits = 3)

write.table(corr_potential_trait_confounders_pos_MVMR_summary, "output/tables/C_summ_MVMR_result.csv")

corr_potential_trait_confounders_pos_MVMR_summary <- corr_potential_trait_confounders_pos_MVMR_summary %>% 
  group_by(outcome_ID) %>%
  mutate(corr_due_to_confounding_MVMR_summ = sum(corr_due_to_confounding_MVMR)) %>%
  mutate(corr_due_to_confounding_summ = sum(corr_due_to_confounding))

```

We compared the difference in phenotypic estimate and MR-estimates to the maximum C for each trait, as shown in the figure below: 

```{r max_C_plot, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), message=FALSE, fig.height=10, dpi=600}

max_C_data <- corr_potential_trait_confounders_pos %>% filter(sig_confounder) %>% 
  #arrange(-corr_due_to_confounding_ratio) %>%
  mutate(total_potential_confounders = n()) %>%
  #slice_head(n = 1) %>%
  left_join(couple_MR_vs_trait_corr %>% 
              dplyr::rename(outcome_ID_IVW_beta = IVW_beta) %>%
              dplyr::rename(outcome_ID_IVW_se = IVW_se) %>%
              dplyr::rename(outcome_ID_couple_r = couple_r) %>% 
              dplyr::rename(outcome_ID_couple_r_se = couple_r_se) %>% 
              dplyr::select(outcome_ID, outcome_ID_IVW_beta, outcome_ID_IVW_se, outcome_ID_couple_r, outcome_ID_couple_r_se)) %>%
  mutate(same_trait_corr_minus_MR_est = abs(outcome_ID_couple_r)- abs(outcome_ID_IVW_beta)) %>%
  dplyr::rename(Z_beta = IVW_beta) %>%
  dplyr::rename(Z_se = IVW_se) %>%
  dplyr::rename(Z_pval = IVW_pval) %>%
  mutate(Z_sq_beta = Z_beta^2) %>%
  mutate(Z_sq_se = sqrt(4*Z_beta^2*Z_se^2+2*Z_se^4)) %>%
  dplyr::rename(Z_AM_beta = exposure_ID_AM_IVW_beta) %>%
  dplyr::rename(Z_AM_se = exposure_ID_AM_IVW_se) %>%
  mutate(Z_prod = Z_sq_beta*Z_AM_beta) %>%
  mutate(Z_prod_var = variance_of_product(Z_sq_beta, Z_sq_se, Z_AM_beta, Z_AM_se)) %>%
  mutate(Z_prod_se = sqrt(Z_prod_var)) %>%
  group_by(outcome_ID) %>% 
  mutate(corr_due_to_confounding_summ = sum(corr_due_to_confounding)) %>%
  mutate(corr_due_to_confounding_summ_SE = sqrt(sum(Z_prod_var))) %>%
  mutate(same_trait_corr_minus_MR_est_se = sqrt(variance_of_sum(outcome_ID_IVW_se, outcome_ID_couple_r_se))) %>%
  mutate(Phenotype = outcome_description) %>% 
  mutate(Phenotype = case_when(Phenotype == "Basal metabolic rate" ~ "BMR",
                               Phenotype == "Systolic blood pressure, automated reading" ~ "SBP",
                               Phenotype == "Pulse rate, automated reading" ~ "Pulse rate",
                               Phenotype == "Body fat percentage" ~ "BFP",
                               Phenotype == "Standing height" ~ "Height",
                               Phenotype == "Reticulocyte percentage" ~ "Reticulocyte %",
      
                               #Phenotype == "Weight" & outcome_ID == "21002_irnt" ~ "Weight",
                               Phenotype == "Forced vital capacity (FVC), Best measure" ~ "FVC",
                               Phenotype == "Red blood cell (erythrocyte) count" ~ "RBC count",
                               Phenotype == "Place of birth in UK - north co-ordinate" ~ "Place of birth, NC",
                               TRUE ~ Phenotype)) %>%
  mutate(label_phenos = case_when(outcome_description %in% interesting_phenotypes ~ TRUE, TRUE ~ FALSE)) %>% 
  dplyr::select(Phenotype, label_phenos, same_trait_corr_minus_MR_est, corr_due_to_confounding_summ, 
                same_trait_corr_minus_MR_est_se, corr_due_to_confounding_summ_SE) %>% unique()



max_C_plot <- ggplot2::ggplot(data = max_C_data, 
                              ggplot2::aes(x = corr_due_to_confounding_summ, 
                                           y = same_trait_corr_minus_MR_est, 
                                           label = Phenotype)) +
  
  ggplot2::geom_errorbar(
    ggplot2::aes(ymin = same_trait_corr_minus_MR_est - 1.96*same_trait_corr_minus_MR_est_se, 
                 ymax = same_trait_corr_minus_MR_est + 1.96*same_trait_corr_minus_MR_est_se),
    colour = "grey",
    width = 0
  ) +
  ggplot2::geom_errorbar(
    ggplot2::aes(xmin = corr_due_to_confounding_summ - 1.96*corr_due_to_confounding_summ_SE, 
                 xmax = corr_due_to_confounding_summ + 1.96*corr_due_to_confounding_summ_SE),
    colour = "grey",
    width = 0
  ) +
  
  geom_point(alpha = 1.2) +
  #scale_color_manual(values = custom_col[c(1,2)], labels=c("Non-significant","BF-signifncant difference")) +
  #geom_text(show.legend = FALSE) +
  geom_abline(slope=1, intercept=0) +

  theme_minimal() +
  ggplot2::labs(x = paste0("C-sum"),
                y = paste("Correlation - MR-estimate")) +
  theme(legend.title=element_blank()) +
  geom_text_repel(data = subset(max_C_data, label_phenos==TRUE),
                  nudge_y = -0.25 - subset(max_C_data, label_phenos==TRUE)$same_trait_corr_minus_MR_est, size = 3.25,
                  direction = "x", show.legend = FALSE, box.padding = 0.5)


max_C_plot

## combine `max_C_plot` with `corr_vs_MR_effects_plot`

corr_vs_MR_comb <- plot_grid(plotlist = list(corr_vs_MR_effects_plot, max_C_plot), nrow=2, ncol = 1, byrow = T, labels="AUTO")

corr_vs_MR_comb

```

Lastly, we attempted to compute a MVMR sum of the effects of all potential confounders as shown in the DAG below and compared these C-statistics to the difference between phenotypic correlation and MR-estimates, however the estimates did not make sense so we decided not to pursue this analysis.

```{r MVMR_C_model, echo = FALSE}
library(knitr)
knitr::include_graphics("assets/MVMR_confounder_model.png", error = FALSE)
```

```{r MVMR_C_plot, echo = FALSE, eval = FALSE, message=FALSE}

discrepentant_traits <- unique(corr_potential_trait_confounders_pos$outcome_ID)

output <- numeric()
for(i in 1:length(potential_trait_confounders_MVMR)){
  
  outcome_ID <- discrepentant_traits[i]

  corr_potential_trait_confounders_pos_i <- corr_potential_trait_confounders_pos %>% filter(outcome_ID == !!outcome_ID)
  
  MVMR_result_org <- potential_trait_confounders_MVMR[[i]]
  
  
  MVMR_result_org <- MVMR_result_org %>% mutate(beta_2_var = 4*beta^2*se^2+2*se^4) %>%
    mutate(beta_2_se = sqrt(beta_2_var)) %>% 
    left_join(corr_potential_trait_confounders_pos_i %>% dplyr::select(exposure_ID, exposure_ID_AM_IVW_beta, exposure_ID_AM_IVW_se), by = c("exposure" = "exposure_ID")) %>%
    dplyr::rename(MVMR_Z_beta = beta) %>%
    dplyr::rename(MVMR_Z_se = se) %>%
    dplyr::rename(MVMR_Z_pval = pval) %>%
    mutate(MVMR_Z_sq_beta = MVMR_Z_beta^2) %>%
    dplyr::rename(MVMR_Z_sq_se = beta_2_se) %>% 
    dplyr::rename(Z_AM_beta = exposure_ID_AM_IVW_beta) %>%
    dplyr::rename(Z_AM_se = exposure_ID_AM_IVW_se) %>%
    mutate(Z_prod = MVMR_Z_sq_beta*Z_AM_beta) %>%
    mutate(Z_prod_var = variance_of_product(MVMR_Z_sq_beta, MVMR_Z_sq_se, Z_AM_beta, Z_AM_se)) 
    
  MVMR_result_nom <- MVMR_result_org %>% filter(MVMR_Z_pval < 0.05)
  MVMR_result_BF <- MVMR_result_org %>% filter(MVMR_Z_pval < 0.05/dim(MVMR_result_org)[1])
  
  num_confounders_nom <- dim(MVMR_result_org)[1]
  num_confounders_BF <- dim(MVMR_result_BF)[1]
  
  row <- cbind(outcome_ID)
  for(filter_type in c("nom", "BF")){
    
    if(filter_type == "nom"){
      t <- MVMR_result_org %>% filter(MVMR_Z_pval < 0.05)
    }
    if(filter_type == "BF"){
      t <- MVMR_result_org %>% filter(MVMR_Z_pval < 0.05/dim(MVMR_result_org)[1])
    }
    
    num_confounders <- dim(t)[1]
    C_sum <- sum(t$Z_prod)
    C_sum_var <- sum(t$Z_prod_var)
    C_sum_se <- sqrt(C_sum_var)
    
    temp_row <- cbind(num_confounders, C_sum, C_sum_var, C_sum_se)
    colnames(temp_row) <- paste0(c("num_confounders", "C_sum", "C_sum_var", "C_sum_se"), "_", filter_type)
    row <- cbind(row, temp_row)
    
  }

  
  output <- rbind(output, row)

}

output <- output %>% as_tibble() %>% type_convert() %>% mutate(C_sum_pval = 2*pnorm(-abs(C_sum_nom/C_sum_se_nom)))

MVMR_C_data <- max_C_data %>% left_join(output)

MVMR_C_plot <- ggplot2::ggplot(data = MVMR_C_data, ggplot2::aes(x = C_sum_nom ,
                                                       y =  same_trait_corr_minus_MR_est)) +
  geom_point(alpha = 1.2) +
  #scale_color_manual(values = custom_col[c(1,2)], labels=c("Non-significant","BF-signifncant difference")) +
  #geom_text(show.legend = FALSE) +
  #geom_abline(slope=1, intercept=0) +

  # ggplot2::geom_errorbar(
  #   ggplot2::aes(ymin = MR_est - IVW_se, ymax = MR_est + IVW_se),
  #   colour = "grey",
  #   width = 0
  # ) +
  # ggplot2::geom_errorbar(
  #   ggplot2::aes(xmin = Correlation - couple_r_se, xmax = Correlation + couple_r_se),
  #   colour = "grey",
  #   width = 0
  # ) +

  theme_minimal() +
  ggplot2::labs(x = paste0("MVMR C-sum (p < 0.05)"),
                y = paste("abs(correlation) - abs(MR-estimate)")) +
  theme(legend.title=element_blank()) +
  geom_abline(slope=1, intercept=0)


MVMR_C_plot


```

# Single-trait causal effects in couples. 

```{r AM_data_prep, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE, include = FALSE}

AM_result_sig <-
  household_MR_summary_AM %>% filter(IVW_pval < 0.05 / num_tests_by_PCs)
AM_sig_traits <-
  household_MR_summary_AM %>% filter(IVW_pval < 0.05 / num_tests_by_PCs) %>% pull(exposure_ID) %>% unique()
AM_sex_spec_result_sig <-
  household_MR_binned_het_sex_spec %>% filter(same_trait == TRUE) %>% filter(exposure_ID %in% AM_sig_traits) %>% filter(sex_het_pval < 0.05 /
                                                                                                                          num_tests_by_PCs_AM_sig)

```

- For each trait in the pipeline, we tested the causal effect within couples (i.e. the number of analyses corresponds to the number of traits (`r dim(traits_final)[1]`)). 
- The significant results after adjusting for multiple hypothesis testing are shown below. 
- `IVW_meta_beta` and `IVW_meta_pval` corresponds to the beta and p-value of the meta-analyzed MR across sexes, respectively (i.e. MR estimates were computed in each sex-seperately using sex-specific SNP-exposure and SNP-outcome results and then meta-analyzed).  
- After adjusting for multiple tests (p < `0.05/`r num_tests_by_PCs``), identified `r dim(AM_result_sig)[1]` significant assortative mating (same-trait) MR results (corresponding to the table below and table "BF_sig_same_trait_MR" in the excel document).


```{r, AM_sig, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE, include = TRUE, message = FALSE}

AM_result_sig.DT <- AM_result_sig %>% dplyr::select(exposure_description, exposure_ID, IVW_beta, IVW_se, IVW_pval) %>% unique() %>% 
  dplyr::rename(trait_description = exposure_description) %>%
  dplyr::rename(trait_ID = exposure_ID)
  
DT::datatable(AM_result_sig.DT, width = "100%") %>% formatSignif(columns=c('IVW_beta', 'IVW_se', 'IVW_pval'), digits=3)

AM_sig_sex_het_nom <- household_MR_binned_het_sex_spec %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% filter(sex_het_pval < 0.05)
AM_sig_sex_het_nom <- replace_Neale_ID(AM_sig_sex_het_nom, traits_corr2_filled,"exposure_ID", "exposure_description") %>%
  dplyr::rename(trait_description = exposure_description) %>%
  dplyr::rename(trait_ID = exposure_ID) %>%
  dplyr::select(trait_description, trait_ID, everything())

AM_result_sig_nom.DT <- AM_sig_sex_het_nom %>% dplyr::select(trait_description, trait_ID, exposure_sex, IVW_beta, IVW_se, IVW_pval, sex_het_pval) %>% unique()


AM_sensitivity <- bind_rows(household_MR_SNPmeta_sensitivity) %>% filter(same_trait)
AM_joint.DT <- left_join(AM_result_sig, AM_sensitivity) %>% 
  dplyr::rename(trait_description = exposure_description) %>%
  dplyr::rename(trait_ID = exposure_ID) %>%
  dplyr::select(trait_description, trait_ID, wt_median_summary, wt_median_pval, wt_mode_summary, wt_mode_pval)

non_sig_MR_mode <- AM_joint.DT %>% filter(wt_mode_pval > 0.05)
non_sig_MR_median <- AM_joint.DT %>% filter(wt_median_pval > 0.05)

```

## Heterogeneity stats.

- In the single trait analysis, we also decided to investigate Cochran’s heterogeneity Q-stat to identify traits with high heterogeneity.
- This would indicate that associations between the index genotype and partner’s phenotype may not only act indirectly through causal relationship between the traits, but some direct effect is present. 
- Ideally, for the majority of the traits, the heterogeneity would be low, confirming that (despite what Tenesa suggests) there are no direct index genome to partner phenome effects.
- We tested heterogeneity statistics using the function [mr_heterogeneity](https://mrcieu.github.io/TwoSampleMR/reference/mr_heterogeneity.html) from the [TwoSampleMR](https://mrcieu.github.io/TwoSampleMR/index.html) package. 

```{r, het_stats, include = FALSE}
sig_het <- household_MR_summary_AM %>% filter(Het_IVW_Qpval_round < 0.05/num_tests_by_PCs)
```

The number of traits with significant heterogeneity in the MR is: `r dim(sig_het)[1]`.

## Sensitivity follow-up.

Of those traits that had a significant causal-effect among couples by IVW, we tested the MR weighted-mode and MR weighted-median method for consistent findings. 

- We found `r dim(non_sig_MR_mode)[1]` trait(s) which was/were not significant using the weighted-mode approach (p > 0.05). 
- We found `r dim(non_sig_MR_median)[1]` trait(s) which was/were not significant using the weighted-mode approach (p > 0.05). 

Results can be seen in the table below. 

```{r, AM_sensitivity, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE, include = TRUE}


DT::datatable(AM_joint.DT, width = "100%") %>% formatSignif(columns=c('wt_median_pval', 'wt_mode_pval'), digits=3)

```

# Effect of sex, age and time together on causal effects in couples.

## Sex heterogeneity.

- For each AM MR ($X_{p} \sim X_i$), analyses were run in each sex separately. 
- Of the `r dim(AM_sig_traits)[1]` significant results shown above, we tested to see if there was any difference in AM MR estimates between sexes. 
- A similar adjustment for multiple hypothesis testing was applied as above, resulting in `r num_tests_by_PCs_AM_sig` number of effective tests. 
- After adjusting for multiple hypothesis testing (p < `0.05/`r num_tests_by_PCs_AM_sig``), `r dim(AM_sex_spec_result_sig)[1]` traits showed significant differences amongst sexes.
- The table below shows the *nominally significant* results at `p < 0.05` (of which there are `r dim(AM_result_sig_nom.DT)[1]/2`) (a slight variation of this table corresponds to "nominally_sig_sex_diff" in the excel document).
- The $p_{binomial}$ can be calculated as `1 – pbinom(15, 64, 0.05)`, and the fold enrichment as `15/(64*0.05)`.
- Found that female estimates are on average larger than males (result of the paired t-test shown below).

```{r AM_sex_het_nom, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE}

DT::datatable(AM_result_sig_nom.DT, width = "100%") %>% formatSignif(columns=c('IVW_beta', 'IVW_pval', 'sex_het_pval'), digits=3)

nominally_sig_sex_differences <- AM_result_sig_nom.DT %>% pivot_wider(names_from = exposure_sex, values_from = c(IVW_beta, IVW_se, IVW_pval)) 

t.test(nominally_sig_sex_differences$IVW_beta_male, nominally_sig_sex_differences$IVW_beta_female, paired = T)

```

## MR binned results (exploring effect of age and time-together).

```{r binned_results_prep, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE, include = FALSE}


slope_sig_age <- household_MR_binned_het_SNPmeta %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% 
  filter(grouping_var == "age_even_bins") %>% 
  filter(bin_slope_pval < 0.05/num_tests_by_PCs_AM_sig)

slope_sig_tt <- household_MR_binned_het_SNPmeta %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% 
  filter(grouping_var == "time_together_even_bins") %>% 
  filter(bin_slope_pval < 0.05/num_tests_by_PCs_AM_sig)

slope_wt_sig_age <- household_MR_binned_het_SNPmeta %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% 
  filter(grouping_var == "age_even_bins") %>% 
  filter(bin_slope_pval_wt < 0.05/num_tests_by_PCs_AM_sig)

slope_wt_sig_tt <- household_MR_binned_het_SNPmeta %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% 
  filter(grouping_var == "time_together_even_bins") %>% 
  filter(bin_slope_pval_wt < 0.05/num_tests_by_PCs_AM_sig)

sex_het_sig_age <- household_MR_binned_het_sex_spec %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% 
  filter(grouping_var == "age_even_bins") %>% 
  filter(bin_slope_sex_het_pval < 0.05/num_tests_by_PCs_AM_sig) 

sex_het_sig_tt <- household_MR_binned_het_sex_spec %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% 
  filter(grouping_var == "time_together_even_bins") %>% 
  filter(bin_slope_sex_het_pval < 0.05/num_tests_by_PCs_AM_sig)

sex_het_wt_sig_age <- household_MR_binned_het_sex_spec %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% 
  filter(grouping_var == "age_even_bins") %>% 
  filter(bin_slope_sex_het_pval_wt < 0.05/num_tests_by_PCs_AM_sig)

sex_het_wt_sig_tt <- household_MR_binned_het_sex_spec %>% filter(same_trait) %>% filter(exposure_ID %in% AM_sig_traits) %>% 
  filter(grouping_var == "time_together_even_bins") %>% 
  filter(bin_slope_sex_het_pval_wt < 0.05/num_tests_by_PCs_AM_sig)

```

- For each AM MR ($X_{p} \sim X_i$), analyses were run in the full sample as well as in 5 roughly equal sized bins according to time couple had been together (`time_together_even_bins`) estimated using [time at household variable](https://biobank.ctsu.ox.ac.uk/showcase/field.cgi?id=699), and median age (`age_even_bins`). 
- Of the `r length(AM_sig_traits)` significant results shown above, we tested to see if there was any significant difference in AM MR estimates amongst the two grouping variables. 
- Specifically, the outcome data set was split into 5 bins, and the SNP-outcome effect was estimated in each bin separately (and each sex seperately). 
- These outcome-SNP effects were then used to generate bin-specific MR estimates, using the same SNP-exposure effects from Neale. 
- To estimate the difference among bins, we used two approaches: 
    1. The Cochran's Q test to test for heterogeneity in meta-analyses *(of which we obtained no significant results, so not presented in paper)*.
    2. Tested the significance of the slope of linear model of median bin (either `age` or `time together`) versus the bin-specific MR estimate.  

------------------------------------------------------------------------

Slopes were calculated by estimating the beta-coefficient of a linear regression between MR estimate within each bin (dependent variable) and median bin (independent variable, either age or time at same household). Linear models were run both *unweighted* and *weighted* for the by the inverse of the SE of the MR estimate. 

The number of significant trends ($\beta$ estimates from the model: $\alpha_{bin} \sim median_{bin}$) significant after multiple hypothesis testing (p < `0.05/`r num_tests_by_PCs_AM_sig` = `r 0.05/num_tests_by_PCs_AM_sig``) in each group was:

  - Binned by median *age* of couples, number of significant results after adjusting for number of tests:
      - $\beta$ unweighted, meta-anlayzed across sexes: `r dim(slope_sig_age)[1]`
      - $\beta$ weighted by inverse of the SE, meta-anlayzed across sexes: `r dim(slope_wt_sig_age)[1]`
      - $\beta$ unweighted, significantly different between sexes: `r dim(sex_het_sig_age)[1]/2`
      - $\beta$ weighted by inverse of the SE, significantly different between sexes: `r dim(sex_het_wt_sig_age)[1]/2`
      
  - Binned by *time* couple had been together, number of significant results after adjusting for number of tests:
      - $\beta$ unweighted, meta-anlayzed across sexes: `r dim(slope_sig_tt)[1]`
      - $\beta$ weighted by inverse of the SE, meta-anlayzed across sexes: `r dim(slope_wt_sig_tt)[1]`
      - $\beta$ unweighted, significantly different between sexes: `r dim(sex_het_sig_tt)[1]/2`.
      - $\beta$ weighted by inverse of the SE, significantly different between sexes: `r dim(sex_het_wt_sig_tt)[1]/2`

The significant traits correspond to the following traits: 

```{r sig_sex_diff, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE, include = TRUE}

sex_het_wt_sig_age_traits <- replace_Neale_ID(sex_het_wt_sig_age, traits_corr2_filled, "exposure_ID", "outcome_description") %>% pull(outcome_description) %>% unique()


sex_het_sig_age_sub <- sex_het_sig_age %>% dplyr::select(exposure_ID, exposure_sex, grouping_var, bin_slope_beta, bin_slope_se, bin_slope_pval, bin_slope_sex_het_pval)
sex_het_sig_tt_sub <- sex_het_sig_tt %>% dplyr::select(exposure_ID, exposure_sex, grouping_var, bin_slope_beta, bin_slope_se, bin_slope_pval, bin_slope_sex_het_pval)

sig_sex_diff_dat <- replace_Neale_ID(rbind(sex_het_sig_age_sub, sex_het_sig_tt_sub), traits_corr2_filled, "exposure_ID", "exposure_description")

DT::datatable(sig_sex_diff_dat, width = "100%") %>% formatSignif(columns=c('bin_slope_beta', 'bin_slope_se', 'bin_slope_pval', 'bin_slope_sex_het_pval'), digits=3)

```

A few potential figures we could include for a given trait are shown below. 

```{r binned_figures, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE, include = TRUE}

tar_load(exposure_info)

branches_to_pull <- numeric()
for(descriptions in unique(sig_sex_diff_dat$exposure_description))
{
  
  for(i in 1:length(exposure_info)){
  temp <- exposure_info[[i]]
  if(temp[which(temp$Value=="description"), "Info"] == descriptions) {break}
  }
  branches_to_pull <- c(i, branches_to_pull)
}


figs <- list()

for(i in branches_to_pull){
  
  name <- paste0("exposure_", i)
  figs[[name]] <- tar_read(household_MR_binned_AM_figs, branches = i)[[1]]

}

figs[[1]][[4]]
figs[[1]][[5]]
figs[[2]][[4]]
figs[[2]][[5]]

```

## Phenotypic binned results (exploring effect of age and time-together).

We also tested the phenotypic correlation among the different bins. Below are the traits that show a significant trend either with Pearson correlation (p < `r 0.05/num_tests_by_PCs`) (the table below corresponds to "pheno_binned_corrs" in the excel document). 

```{r, pheno_binned_result, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE}

binned_pheno_corr_summary <- bind_rows(lapply(binned_pheno_corrs, function(x) {x[[2]]}))

pheno_corr_pearson_sig <- binned_pheno_corr_summary %>% filter(bin_slope_pval_pearson < 0.05/num_tests_by_PCs) %>% arrange(desc(group))

pheno_corr_pearson_sig <- replace_Neale_ID(pheno_corr_pearson_sig, traits_corr2_filled, "trait", "trait_description") %>%
  dplyr::rename(trait_ID = trait) %>%
  dplyr::select(trait_description, trait_ID, everything())

## create panel figure 
pheno_corr_pearson_sig$branches_to_pull <- NA

for(i in 1:dim(pheno_corr_pearson_sig)[1]){
  
  trait <- pheno_corr_pearson_sig[i, "trait_ID"][[1]]
  for(k in 1:length(exposure_info)){
  temp <- exposure_info[[k]]
  if(temp[which(temp$Value=="trait_ID"), "Info"] == trait) {break}
  }

  pheno_corr_pearson_sig$branches_to_pull[i] <- k
}


figs <- list()

for(i in 1:dim(pheno_corr_pearson_sig)[1]){
  
  k <- pheno_corr_pearson_sig$branches_to_pull[i] 
  binned_pheno_corrs_i <- tar_read(binned_pheno_corrs, branches = k)[[1]]
  exposure_info_i <- tar_read(exposure_info, branches = k)[[1]]
  pheno_figs <- create_binned_pheno_figs (binned_pheno_corrs_i, exposure_info_i, grouping_var)
  
  name <- paste0("exposure_", i)
  group <- pheno_corr_pearson_sig[i, "group"][[1]]
  fig_num <- ifelse(group == "age_even_bins", 1, 2)
  figs[[name]] <- pheno_figs[[fig_num]] #tar_read(binned_pheno_figs, branches = i)[[1]][[fig_num]]

}

pheno_binned_figures <- plot_grid(plotlist = figs, nrow=2, ncol = 2, byrow = T, labels="AUTO")

## format table 
pheno_corr_pearson_sig.DT <- pheno_corr_pearson_sig %>% dplyr::select(-branches_to_pull) %>%
  mutate(bin_type = case_when(group == "age_even_bins" ~ "age", 
                              group == "time_together_even_bins" ~ "time_together")) %>%
  dplyr::select(-group) %>% dplyr::select(trait_description, trait_ID, bin_type, everything())


format_round_cols <- colnames(dplyr::select_if(pheno_corr_pearson_sig.DT, is.numeric))

DT::datatable(pheno_corr_pearson_sig.DT, width = "100%") %>% formatRound(columns=format_round_cols, digits = 3)


```

```{r, pheno_binned_figs, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE, fig.height=7, dpi=600}

pheno_binned_figures

```

# Impact of confounders on couple correlation.

We tested the impact of confounding on genetic PCs, N/E coordinates and the following traits (however only presenting a selection in paper): 

```{r confounder_traits, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE}
confounder_trait_tibble <- as_tibble(confounder_traits)
confounder_trait_tibble_descript <- replace_Neale_ID(confounder_trait_tibble, traits_corr2_filled, "value", "outcome_description")
confounder_trait_tibble_descript %>% 
  kbl(col.names = NULL) %>%
  kable_styling()
```

## Ratio of confounding to correlation. 

We tested the ratio of confounding to correlation for each confounder trait tested. 

```{r corr_due_to_confounding_ratio, eval = FALSE, echo = FALSE}

## originally ran the confounder analysis using correlations only, but switched to use causal effects. 

corr_impact_by_coords_ratio <- corr_impact_by_coords %>% 
  dplyr::select(Neale_pheno_ID, trait_couple_corr, trait_couple_corr_se, corr_due_to_confounding_all, corr_due_to_confounding_all_se) %>%
  unique() %>%
  mutate(confounder_trait = "household_coords") %>%
  mutate(trait_ratio = corr_due_to_confounding_all/trait_couple_corr)

corr_impact_by_PCs_ratio <- corr_impact_by_PCs %>% 
  dplyr::select(Neale_pheno_ID, trait_couple_corr, trait_couple_corr_se, corr_due_to_confounding_all, corr_due_to_confounding_all_se) %>%
  unique() %>%
  mutate(confounder_trait = "PCs") %>%
  mutate(trait_ratio = corr_due_to_confounding_all/trait_couple_corr)

corr_impact_by_traits_ratio <- corr_impact_by_traits %>% dplyr::rename(confounder_trait = trait_interest) %>% 
  group_by(confounder_trait) %>% 
  dplyr::select(trait_j, confounder_trait, trait_j_couple_corr, trait_j_couple_corr_se, 
                corr_due_to_confounding, corr_due_to_confounding_se) %>% 
  mutate(trait_ratio = corr_due_to_confounding/trait_j_couple_corr) 

corr_impact_by_traits_ratio <- replace_Neale_ID(corr_impact_by_traits_ratio, traits_corr2_filled, "confounder_trait", "confounder_trait_description") %>% 
  ungroup() %>%
  dplyr::select(-confounder_trait) %>% dplyr::rename(confounder_trait = confounder_trait_description)

corr_impact_ratios <- rbind(corr_impact_by_PCs_ratio, corr_impact_by_coords_ratio, corr_impact_by_traits_ratio)

corr_impact_ratios_summary <- corr_impact_ratios %>%
  group_by(confounder_trait) %>%
  dplyr::summarise(mean_ratio = mean(trait_ratio), median_ratio = median(trait_ratio), sd_ratio = sd(trait_ratio), number_of_traits = n())

corr_impact_ratios_summary %>%
  kbl() %>%
  kable_styling()
  
```

```{r corr_due_to_confounding_ratio_causal, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE}

confounder_traits_sub <- confounder_traits[c(1,2,5,10,15)]

corr_impact_by_traits_causal_summary <- corr_impact_by_traits_causal %>% group_by(exposure_ID) %>%
   dplyr::summarise(mean_ratio = mean(corr_due_to_confounding_ratio), 
                    median_ratio = median(corr_due_to_confounding_ratio), sd_ratio = sd(corr_due_to_confounding_ratio), 
                    number_of_traits = n()) 

corr_impact_by_traits_causal_summary <- replace_Neale_ID(corr_impact_by_traits_causal_summary, traits_corr2_filled, "exposure_ID", "outcome_description") 

corr_impact_by_traits_causal_summary <- corr_impact_by_traits_causal_summary[which(corr_impact_by_traits_causal_summary$exposure_ID %in% confounder_traits_sub),] %>% 
  dplyr::rename(confounder_trait =  outcome_description) %>% 
  dplyr::rename(confounder_trait_ID =  exposure_ID) %>% 
  dplyr::select(confounder_trait, confounder_trait_ID, everything())

corr_impact_by_PCs_summary <- corr_impact_by_PCs %>% 
  dplyr::select(Neale_pheno_ID, trait_couple_corr, trait_couple_corr_se, corr_due_to_confounding_all, corr_due_to_confounding_all_se) %>%
  unique() %>%
  mutate(confounder_trait = "PCs") %>%
  mutate(trait_ratio = corr_due_to_confounding_all/trait_couple_corr) %>%
  group_by(confounder_trait) %>%
  dplyr::summarise(mean_ratio = mean(trait_ratio), median_ratio = median(trait_ratio), sd_ratio = sd(trait_ratio), number_of_traits = n()) %>%
  mutate(confounder_trait_ID = NA) 


corr_impact_by_coords_summary <- corr_impact_by_traits_causal %>% filter(exposure_ID=="129_irnt" | exposure_ID=="130_irnt") %>% 
  filter(!(outcome_ID=="129_irnt" | outcome_ID=="130_irnt"))  %>%
  group_by(outcome_ID) %>% mutate(corr_due_to_confounding_all = sum(corr_due_to_confounding)) %>%
  mutate(corr_due_to_confounding_all_se = sqrt(sum(corr_due_to_confounding_se^2))) %>%
  dplyr::select(outcome_ID, outcome_ID_couple_r, outcome_ID_couple_r_se, corr_due_to_confounding_all, corr_due_to_confounding_all_se) %>% unique() %>%
  mutate(confounder_trait = "Birth place coordinates") %>%
  mutate(trait_ratio = corr_due_to_confounding_all/outcome_ID_couple_r) %>%
  group_by(confounder_trait) %>%
  dplyr::summarise(mean_ratio = mean(trait_ratio), median_ratio = median(trait_ratio), sd_ratio = sd(trait_ratio), number_of_traits = n()) %>%
  mutate(confounder_trait_ID = NA) 

corr_impact_summary <- rbind(corr_impact_by_traits_causal_summary, corr_impact_by_coords_summary)

corr_impact_summary %>%
  kbl() %>%
  kable_styling()

```

## Plots.

The figure below shows the 4 traits which showed the greatest impact on couple correlation (mean confounding ratio > 0.01).

```{r confounder_plot, echo = FALSE, eval = FALSE, message=FALSE}

confounder_traits_sub <- confounder_traits[c(1,2,4,5)]

geography_plots <- list()


plots <- list()
for(i in 1:length(confounder_traits_sub)){
  
  trait <- confounder_traits_sub[i]
  trait_description <- confounder_trait_tibble_descript[which(confounder_trait_tibble_descript[["value"]]==trait),"outcome_description"][[1]]

  
  if(trait=="738"){trait_description <- "Average total household \nincome before tax"}
  if(trait=="189_irnt"){trait_description <- "Townsend deprivation index \nat recruitment"}
  if(trait=="6160_1"){trait_description <- "Sports club or gym user"}
  
  
  plot_data <- corr_impact_by_traits %>% filter(trait_interest==trait) %>% filter(trait_interest!=trait_j)
  plot_data <- replace_Neale_ID(plot_data, traits_corr2_filled, "trait_j", "Phenotype")
  plot <-
    ggplot2::ggplot(
      data = plot_data,
      ggplot2::aes(x = trait_j_couple_corr,
                   y = corr_due_to_confounding, Phenotype = Phenotype)
    ) +
        
    ggplot2::geom_errorbar(
      ggplot2::aes(ymin = corr_due_to_confounding - 1.96*corr_due_to_confounding_se, ymax = corr_due_to_confounding + 1.96*corr_due_to_confounding_se),
      colour = "grey",
      width = 0
    ) +
    ggplot2::geom_errorbar(
      ggplot2::aes(xmin = trait_j_couple_corr - 1.96*trait_j_couple_corr_se, xmax = trait_j_couple_corr + 1.96*trait_j_couple_corr_se),
      colour = "grey",
      width = 0
    ) +
    
    theme(plot.title = element_text(hjust = 0.5)) +
    geom_point(alpha = 3 / 4) +

    
    geom_abline(slope=1, intercept=0) +
    coord_cartesian(ylim = c(0, NA)) +
    theme_minimal() +
    ggplot2::labs(
      x = paste0("Couple trait correlation"),
      y = paste0("Couple correlation \ndue to confounding"),
      title = paste0(trait_description)
      #title = paste0("Impact of ", trait_description, "\n on couple correlation")
    ) +
    xlim(0, 1.0) +
    xlim(0, 1.0) +
    
    theme(plot.title = element_text(size=11, hjust = 0.5)) + theme(axis.title = element_text(size = 10))  
    
  plots[[i]] <- plot 
}

i <- 5 

figures_grid <- plot_grid(plotlist = plots, nrow=2, ncol = 2, byrow = F, labels="AUTO")
figures_grid
```

```{r PC_confounder_plot, echo = FALSE, eval = TRUE, message=FALSE}

plot_data <- corr_impact_by_PCs %>% dplyr::select(Neale_pheno_ID, trait_couple_corr, corr_due_to_confounding_all, corr_due_to_confounding_all_se, trait_couple_corr_se, trait_couple_corr_se) %>% unique()

plot_PCs <-
  ggplot2::ggplot(
    data = plot_data,
    ggplot2::aes(x = trait_couple_corr,
                 y = corr_due_to_confounding_all)
  ) +
  theme(plot.title = element_text(hjust = 0.5)) +
  
  ggplot2::geom_errorbar(
    ggplot2::aes(ymin = corr_due_to_confounding_all - 1.96*corr_due_to_confounding_all_se, ymax = corr_due_to_confounding_all + 1.96*corr_due_to_confounding_all_se),
    colour = "grey",
    width = 0
  ) +
  ggplot2::geom_errorbar(
    ggplot2::aes(xmin = trait_couple_corr - 1.96*trait_couple_corr_se, xmax = trait_couple_corr + 1.96*trait_couple_corr_se),
    colour = "grey",
    width = 0
  ) +
  
  
  geom_point(alpha = 3 / 4) +
  

  theme_minimal() +
  ggplot2::labs(
    x = paste0("Couple trait correlation"),
    y = paste0("Couple correlation \ndue to confounding"),
    title = paste0("Principal components")
    #title = paste0("Impact of ", trait_description, "\n on couple correlation")
  ) +
  
  xlim(0, 1.0) +
  theme(plot.title = element_text(size=11, hjust = 0.5)) + theme(axis.title = element_text(size = 10))  +
  geom_abline(slope=1, intercept=0) +
  coord_cartesian(ylim = c(0, NA)) 


```

```{r coord_confounder_plot, echo = FALSE, eval = TRUE, message=FALSE, out.width = "200%", dpi=600}

trait_description <- "Place of birth coordinates"

plot_data <- corr_impact_by_traits_causal %>% filter(exposure_ID=="129_irnt" | exposure_ID=="130_irnt") %>% 
  filter(!(outcome_ID=="129_irnt" | outcome_ID=="130_irnt"))  %>%
  group_by(outcome_ID) %>% mutate(corr_due_to_confounding_all = sum(corr_due_to_confounding)) %>%
  mutate(corr_due_to_confounding_all_se = sqrt(sum(corr_due_to_confounding_se^2))) %>%
  dplyr::select(outcome_ID, outcome_ID_couple_r, outcome_ID_couple_r_se, corr_due_to_confounding_all, corr_due_to_confounding_all_se) %>% unique()


plot_data <- replace_Neale_ID(plot_data, traits_corr2_filled, "outcome_ID", "Phenotype")
coord_plot <-
  ggplot2::ggplot(
    data = plot_data,
    ggplot2::aes(x = outcome_ID_couple_r,
                 y = corr_due_to_confounding_all, Phenotype = Phenotype)
  ) +
  theme(plot.title = element_text(hjust = 0.5)) +
  
  
  ggplot2::geom_errorbar(
    ggplot2::aes(ymin = corr_due_to_confounding_all - 1.96*corr_due_to_confounding_all_se, ymax = corr_due_to_confounding_all + 1.96*corr_due_to_confounding_all_se),
    colour = "grey",
    width = 0
  ) +
  ggplot2::geom_errorbar(
    ggplot2::aes(xmin = outcome_ID_couple_r - 1.96*outcome_ID_couple_r_se, xmax = outcome_ID_couple_r + 1.96*outcome_ID_couple_r_se),
    colour = "grey",
    width = 0
  ) +
  
  geom_point(alpha = 3 / 4) +
  
  theme_minimal() +
  ggplot2::labs(
    x = paste0("Couple trait correlation"),
    y = paste0("Couple correlation \ndue to confounding"),
    title = paste0(trait_description)
    #title = paste0("Impact of ", trait_description, "\n on couple correlation")
  ) +
  xlim(0, 1.0) +
  theme(plot.title = element_text(size=11, hjust = 0.5)) + theme(axis.title = element_text(size = 10))  +
  geom_abline(slope=1, intercept=0) +
  coord_cartesian(ylim = c(0, NA)) 


```

```{r causal_confounder_plot, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), message=FALSE, dpi=600, fig.height=7}

confounder_traits_sub <- confounder_traits[c(1,2,5,10,15)]


plots <- list()
for(i in 1:length(confounder_traits_sub)){
  
  trait <- confounder_traits_sub[i]
  trait_description <- confounder_trait_tibble_descript[which(confounder_trait_tibble_descript[["value"]]==trait),"outcome_description"][[1]]

  
  
  if(trait=="738"){trait_description <- "Average total household \nincome before tax"}
  if(trait=="189_irnt"){trait_description <- "Townsend deprivation index \nat recruitment"}
  if(trait=="6160_1"){trait_description <- "Sports club or gym user"}
  
  
  plot_data <- corr_impact_by_traits_causal %>% filter(exposure_ID==trait) %>% filter(exposure_ID!=outcome_ID)
  if(dim(plot_data)[1]==0) next
  
  plot_data <- replace_Neale_ID(plot_data, traits_corr2_filled, "outcome_ID", "Phenotype")
  plot <-
    ggplot2::ggplot(
      data = plot_data,
      ggplot2::aes(x = outcome_ID_couple_r,
                   y = corr_due_to_confounding, Phenotype = Phenotype)
    ) +
    theme(plot.title = element_text(hjust = 0.5)) +
    
        
    ggplot2::geom_errorbar(
      ggplot2::aes(ymin = corr_due_to_confounding - 1.96*corr_due_to_confounding_se, ymax = corr_due_to_confounding + 1.96*corr_due_to_confounding_se),
      colour = "grey",
      width = 0
    ) +
    ggplot2::geom_errorbar(
      ggplot2::aes(xmin = outcome_ID_couple_r - 1.96*outcome_ID_couple_r_se, xmax = outcome_ID_couple_r + 1.96*outcome_ID_couple_r_se),
      colour = "grey",
      width = 0
    ) +
    
    geom_point(alpha = 3 / 4) +
    
    theme_minimal() +
    ggplot2::labs(
      x = paste0("Couple trait correlation"),
      y = paste0("Couple correlation \ndue to confounding"),
      title = paste0(trait_description)
      #title = paste0("Impact of ", trait_description, "\n on couple correlation")
    ) +
    xlim(0, 1.0) +
    theme(plot.title = element_text(size=11, hjust = 0.5)) + theme(axis.title = element_text(size = 10))  +
    geom_abline(slope=1, intercept=0) +
    coord_cartesian(ylim = c(0, NA)) 
  
  plots[[i]] <- plot 
}

i <- 6
plots[[i]] <- coord_plot 
figures_grid <- plot_grid(plotlist = plots, nrow=3, ncol = 2, byrow = F, labels="AUTO")
figures_grid


```

# Comparison of paths from index to partner.

As a recap, the model is below. 

```{r YiYp_adj_model, echo = FALSE}
library(knitr)
knitr::include_graphics("assets/two_trait_model_yiyp_adj.png", error = FALSE)
```

We sought to compare the different paths from $X_i$ to $Y_p$ using the estimates computed as $\omega$, $\rho$ and $\gamma$. 

We considered the regression of the following estimates vs $\omega$ with an intercept passing through the origin (i.e. $estimate \sim \omega + 0$:

1. $\rho$
2. $\gamma$
3. $\rho$ + $\gamma$
4. $\rho_{resid} + \gamma$, where $\rho$  is residualized for $\gamma$ (i.e. $\rho_{resid} = resid(lm(\rho \sim \gamma))$) 

The results are summarized below (and provided in table "omega_regression_estimates" in the excel document): 

```{r regression_summary, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE)}

data <- proxyMR_comparison_summary_yiyp_adj_SNPmeta

data <- data %>%
  mutate(
    rho_larger = case_when(
      gam_vs_rho_BF_sig == TRUE & rho_beta > gam_beta ~ TRUE,
      gam_vs_rho_BF_sig == TRUE &
        rho_beta < gam_beta ~ FALSE,
      gam_vs_rho_BF_sig == FALSE ~ NA
    )
  ) %>%
  mutate(omega_larger = case_when((omega_beta - rho_beta - gam_beta) > max(rho_beta, gam_beta) ~ TRUE, 
                                  TRUE ~ FALSE))
  

                
                
data$rho_resid <- resid(lm(rho_beta ~ gam_beta, data = data))
data$gam_rho_resid <- data$rho_resid + data$gam_beta

result <- numeric()
for(outcome in c("rho", "gam", "gam_rho", "gam_rho_resid")){
  
  col_to_rename <- paste0(outcome, "_beta")
  if(outcome=="gam_rho_resid") {col_to_rename <- outcome}
  data_model <- data %>% dplyr::rename(outcome = !!col_to_rename)
  model <- summary(lm(outcome ~ omega_beta + 0, data = data_model)) 
  r2 <- model %>% glance %>% pull(r.squared)
  model_summary <- model %>% tidy() %>% 
    mutate(term = "omega") %>% mutate(outcome = outcome) %>% mutate(r.squared = r2)
  
  
  result <- rbind(result, model_summary)
  
}

omega_regression_estimates <- result %>% dplyr::select(outcome, everything()) %>% filter(outcome !="gam_rho")

omega_regression_estimates %>%
  kbl() %>%
  kable_styling()


```

Of note, we found `r  table(data$gam_vs_rho_BF_sig)[[2]]` relationships which were significantly different between $\rho$ and  $\gamma$, whereby `r table(data$rho_larger)[[2]]` (`r (table(data$rho_larger)[[2]]/table(data$gam_vs_rho_BF_sig)[[2]])*100`%) showed larger effects through $\rho$ and the other `r table(data$rho_larger)[[1]]` (`r (table(data$rho_larger)[[1]]/table(data$gam_vs_rho_BF_sig)[[2]])*100`%) showed larger effects through $\gamma$. 

```{r remove_opp_sign, include = FALSE}

opp_sign_rows <- numeric()
for(col1 in c("gam_beta", "rho_beta", "gam_rho_resid", "omega_beta")){
  
  for(col2 in c("gam_beta", "rho_beta", "gam_rho_resid", "omega_beta")){
  
      temp <- which(sign(data[[col1]])!=sign(data[[col2]]))
      opp_sign_rows <- c(opp_sign_rows, temp)
  }
  

}

opp_sign_rows <- unique(opp_sign_rows)

```

To further explore these estimates, performed paired t-tests to find the significant differences between pairs of estimates. We first removed all instances where the sign did not match between any two pairings. In total, of the `r dim(data)[1]` trait pairs under consideration, `r length(opp_sign_rows)` were removed. 

```{r boxplot, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), message=FALSE, warning = FALSE}

df <- data[-opp_sign_rows,] %>% dplyr::select(exposure_ID, outcome_ID, gam_beta, rho_beta, gam_rho_resid, omega_beta)

data_boxplot  <- df %>%
  pivot_longer(cols = !c("exposure_ID", "outcome_ID"), names_to = "data", values_to = "beta") %>%
  mutate_at("data", as.factor) %>%
  mutate(
    data = fct_recode(as.factor(data),
                      Gamma = "gam_beta",
                      Rho = "rho_beta",
                      Gamma_Rho_resid = "gam_rho_resid",
                      Omega = "omega_beta")
  )


cmpr <- list(c("Omega","Rho"), c("Gamma","Rho"), c("Rho", "Gamma_Rho_resid"))

position <- data_boxplot %>% group_by(data) %>% dplyr::summarise(median=median(beta)) %>% 
  arrange(-desc(median)) %>% pull(data)

data_boxplot$data <- factor(data_boxplot$data,levels=position)

data_boxplot <- data_boxplot %>% filter(data != "Gamma_Rho")
pairwise.t.test(data_boxplot$beta, data_boxplot$data, p.adjust = "none", paired = T)

boxplot <- data_boxplot %>%
  ggplot( aes(x=data, y=beta)) +
  geom_boxplot(outlier.shape = NA) +

  stat_compare_means(data = data_boxplot, comparisons = cmpr, label.y = c(0.6, 0.4, 0.5), method = "t.test" , paired =T) + #,
                     # label = "p.signif",
                     # symnum.args = list(cutpoints = c(0, 0.05, 1),
                     #                    symbols = c("*", "ns"))) +



  coord_cartesian(ylim=c(-0.5, 0.75)) +
  scale_fill_viridis(discrete = TRUE, alpha=0.6, option="A") +
  theme_ipsum() +
  theme(
    legend.position="none",
    plot.title = element_text(size=11)
  ) +
  #labs(x ="Estimate", y = "") +
  scale_x_discrete(labels=c("Omega" = "\u03C9", 
                            "Gamma" = "\u03B3", "Gamma_Rho_resid" = expression( paste(rho ["resid"], " + ", gamma)), "Rho" = "\u03C1")) +
  theme(axis.title.x=element_blank(),
        axis.title.y=element_blank()) 


```


```{r xy_boxplot_combined, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), message=FALSE, warning = FALSE, fig.height = 10}

figures_grid <- plot_grid(plotlist = list(proxyMR_prod_comparison_fig_yiyp_adj, boxplot), nrow=2, ncol = 1, byrow = T, labels = c('', 'E'))
figures_grid

```

## Paths summary.

A summary of the $\omega$, $\rho$, $\gamma$ and $\rho_{resid} + \gamma$ estimates are summarized in the table below. The differences between the estimates (according to a Z-test) are given in the last three columns. Trait pairs were pruned such that:  if there is a pair A-B and C-D and if [`max(corr(A,C)*corr(B,D),corr(A,D)*corr(B,C))`] > `r corr_trait_threshold` then one pair is dropped. Pairs were prioritized based on the $\omega$ p-value (this table corresponds to "paths_summary" in the excel document). 


```{r paths_summary, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE}

tar_load(proxyMR_comparison_summary_yiyp_adj_SNPmeta_prune)

data <- proxyMR_comparison_summary_yiyp_adj_SNPmeta_prune
data$rho_resid <- resid(lm(rho_beta ~ gam_beta, data = data))
data$gam_rho_resid <- data$rho_resid + data$gam_beta


paths_summary <- data %>% dplyr::select(exposure_description, outcome_description, gam_beta, rho_beta, omega_beta, gam_rho_resid, gam_vs_rho_pval, omega_vs_rho_pval, omega_vs_gam_pval, omega_vs_gam_rho_resid_pval)

format_round_cols <- c(colnames(dplyr::select(paths_summary, ends_with("_beta"))), "gam_rho_resid")
format_signif_cols <- colnames(dplyr::select(paths_summary, ends_with("_pval")))


DT::datatable(paths_summary, width = "100%") %>% formatRound(columns=format_round_cols, digits = 3) %>%
  formatSignif(columns = format_signif_cols, digits = 3)

  

```

```{r large_results_table}

traits_description <- traits_final %>% dplyr::select(Neale_pheno_ID, description, variable_type, category, r2, N_pairs, num_IVs_pass_het) %>% as_tibble() %>% 
  dplyr::rename(num_IVs = num_IVs_pass_het) %>%
  dplyr::rename(n_pairs = N_pairs) %>%
  dplyr::rename(pheno_couple_r2 = r2) %>% dplyr::rename(exposure_ID = Neale_pheno_ID)

# changed diff_p name 
MR_result_table <- corr_vs_MR_effects_dat %>% dplyr::select(exposure_ID, Correlation, couple_r_se, n_exposure, n_outcome, MR_est, IVW_se, IVW_pval, diff_p) %>%
  dplyr::rename(pheno_couple_r = Correlation) %>%
  dplyr::rename(pheno_couple_r_se = couple_r_se) %>%
  dplyr::rename(IVW_beta = MR_est) %>%
  dplyr::rename(couple_r_vs_IVW_diff_pval = diff_p) 

AM_sex_spec_result <-
  household_MR_binned_het_sex_spec %>% filter(same_trait == TRUE) %>% dplyr::select(exposure_ID, exposure_sex, IVW_beta, IVW_se, IVW_pval, sex_het_pval) %>% unique()

AM_binned_result <- 
  household_MR_binned_het_SNPmeta %>% filter(same_trait) %>% dplyr::select(exposure_ID, grouping_var, bin_slope_beta_wt, bin_slope_se_wt, bin_slope_pval_wt) %>% 
  mutate(grouping_var_rename = case_when(grouping_var=="age_even_bins" ~ "age",
                                         grouping_var=="time_together_even_bins" ~ "time_together")) %>% dplyr::select(-grouping_var) %>%
  dplyr::rename(bin_slope_beta_MR_wt = bin_slope_beta_wt) %>%
  dplyr::rename(bin_slope_se_MR_wt = bin_slope_se_wt) %>%
  dplyr::rename(bin_slope_pval_MR_wt = bin_slope_pval_wt)

  


sex_differences_summary <- AM_sex_spec_result %>% pivot_wider(names_from = exposure_sex, values_from = c(IVW_beta, IVW_se, IVW_pval)) 
binned_summary <- AM_binned_result %>% pivot_wider(names_from = grouping_var_rename, values_from = c(bin_slope_beta_MR_wt, bin_slope_se_MR_wt, bin_slope_pval_MR_wt)) 

binned_corr_summary <- binned_pheno_corr_summary %>% 
  mutate(grouping_var_rename = case_when(group=="age_even_bins" ~ "age",
                                         group=="time_together_even_bins" ~ "time_together")) %>% unique() %>% dplyr::select(-group) %>%
  pivot_wider(names_from = grouping_var_rename, values_from = c(bin_slope_beta_pearson, bin_slope_se_pearson, 
                                                                bin_slope_pval_pearson,
                                                                bin_slope_beta_spearman, bin_slope_se_spearman, bin_slope_pval_spearman)) %>%
  dplyr::rename(exposure_ID = trait)

df_list <- list(traits_description, MR_result_table, sex_differences_summary, binned_summary, binned_corr_summary)
final_table <- df_list %>% reduce(full_join, by='exposure_ID') %>%
  dplyr::rename(trait_ID = exposure_ID) %>%
  dplyr::rename(trait_description = description) %>%
  dplyr::select(trait_description, trait_ID, everything())
 

```

```{r write_tables, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE), echo = FALSE}

################################# 
## Tables to include in paper ###
#################################

## Cases where correlation is > MR effects: `corr_vs_MR_effects_sig_dat`

## Potential confounders for all cases where corr is > MR effects: `sig_confounders`

## Significant same-trait couple MR effects: `AM_result_sig.DT`

## Nominally significant same-trait sex-difference: `nominally_sig_sex_differences`

## Phenotypic correlations by bin: `binned_pheno_corr_sig.DT`

## Ratio of confounding to correlation: `corr_impact_ratios_summary`

## Omega regression estimates: `omega_regression_estimates`

## Pruend paths summary: `paths_summary`

main_tables_file <- "docs/manuscript_tables/manuscript_tables.xlsx"
supplementary_tables_file <- "docs/manuscript_tables/supplementary_tables.xlsx"

write.xlsx_wrapper <- function(dat, ...){
  
  write.xlsx(as.data.frame(dat), ...)
}

## Main tables 
write.xlsx_wrapper(corr_impact_summary, file=main_tables_file, sheetName="Table 1 corr_impact_ratios_summary", row.names=FALSE)
write.xlsx_wrapper(omega_regression_estimates, file=main_tables_file, sheetName="Table 2 omega_regression_estimates", append=TRUE, row.names=FALSE)

## Supplementary tables
write.xlsx_wrapper(final_table, file=supplementary_tables_file, sheetName="Supp Table 1 complete_summary", row.names=FALSE)
write.xlsx_wrapper(AM_result_sig.DT, file=supplementary_tables_file, sheetName="Supp Table 2 BF_sig_same_trait_MR", append=TRUE, row.names=FALSE)
write.xlsx_wrapper(nominally_sig_sex_differences, file=supplementary_tables_file, sheetName="Supp Table 3 nominally_sig_sex_diff", row.names=FALSE)
write.xlsx_wrapper(pheno_corr_pearson_sig.DT, file=supplementary_tables_file, sheetName="Supp Table 4 pheno_binned_corrs", append=TRUE, row.names=FALSE)
write.xlsx_wrapper(corr_vs_MR_effects_sig_dat, file=supplementary_tables_file, sheetName="Supp Table 5 sig_corr_vs_MR", append = TRUE, row.names=FALSE)
write.xlsx_wrapper(sig_confounders, file=supplementary_tables_file, sheetName="Supp Table 6 sig_confounders", append=TRUE, row.names=FALSE)
write.xlsx_wrapper(paths_summary, file=supplementary_tables_file, sheetName="Supp Table 7 paths_summary", append=TRUE, row.names=FALSE)



```
