---
title: "Update Meeting: 14-07-2021"
author: "Jenny Sjaarda"
date: "2021-06-14"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
  toc: true
---


```{r setup, include = FALSE}

options(scipen=999)
source("code/settings.R")
library(DT)
library(knitr)
library(targets)
```

# Last week summary.

- Update phenotypes, make sure blood chemistry values are included (see list of biomarkers included [here](https://biobank.ndph.ox.ac.uk/ukb/label.cgi?id=17518)).
- Non-urgent analysis: cross partner heritability. 
  - For this we need a full GWAS of index genotype on partner trait (up until this point only run on IVs of interest).
- GWAS of PCs 1-40, adjusted for age and sex, to get IVs (Neale has not run this).
- Run MR in same individual (i.e. standard MR).

# Update.

## Reprocessed PHESANT phenotypes.

To include all possible phenotypes, reprocessed PHESANT for all new phenotype files that have been added to SGG databased since starting the project and added them to the pipeline. The list of phenotypes included can be seen below.

```{r new_phenos, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE)}

tar_load(data_UKBB_directory)
new_files <- c("ukb39955", "ukb42673", "ukb44073", "ukb44073")

data_UKBB_directory_new <- data_UKBB_directory %>% dplyr::filter(file %in% !!new_files)

DT::datatable(data_UKBB_directory_new, width = "100%")

```

## Identified a problem with the Neale summary stats.

- Only affecting the `low_confidence_variant` column. 
- This filter was applied before generating IVs for each summary file that had been downloaded.
- See a description of the issue on [the Neale github issues page](https://github.com/Nealelab/UK_Biobank_GWAS/issues/20).
- Currently in the process of updating the impacted files:
  - Updated Neale GWAS files can be found here: `/data/sgg2/jenny/data/Neale_UKBB_GWAS/neale_files/variant_flag_update`.  
  - Also will recreate the IVs for these files (~70 of our files), under the same name. 
-  As far as I can tell the issue is only in one direction, where some SNPs that were not really low confidence variants (by their definition) were listed as `low_confidence_variant = TRUE`, not the other way around. In other words, the mistake means that we would only be filtering out variants that we should keep.

## Adding PCs to the pipeline. 

- Currently running GWAS of first 40 PCs, residualized for `age` and `sex`. Applied inverse normal rank transformation first.  
- Test correlation between partners. 

```{r pc_corr, echo = FALSE, eval = ifelse(Sys.info()["sysname"]=="Linux", TRUE, FALSE)}

tar_load(PCs_corr)
PCs_corr.DT <- PCs_corr %>% mutate(r = sqrt(as.numeric(r2))) 
DT::datatable(PCs_corr.DT, width = "100%")  %>% formatSignif(columns=c('r2', 'r'), digits=3)

```

# Next steps. 

Define the following: 

**Goal**: To use Mendelian Randomization (MR) to estimate a causal effect of ${X}$ on ${Y}$ between household pairs (i.e. *couples*). 

Definitions: 

- Let ${X}$ and ${Y}$ denote two random variables representing complex traits. 
- Let ${X}_i$ and ${X}_p$ denote ${X}$ in an index case and in their partner. 
- The genotype data of the SNPs to be used as instrumental variables (IVs) is denoted by ${G}$. 
- Let ${G}_i$ and ${G}_p$ denote ${G}$ in an index case and in their partner. 
- To simplify notation we assume that $E(X) = E(Y) = E(G) =  0$ and $Var(X) = Var(Y) = Var(G) = 1$. 
- The effect sizes of the instruments on $X$ is denoted by ${\beta}$. 
- The causal effect is denoted by ${\alpha}$.
- Let us assume the following models.


$$
\begin{align}
(1) \; X_{i} = G_{i}\cdot {\beta_{x, i\rightarrow i} } + \epsilon \\
(2) \; X_{p} = G_{p}\cdot {\beta_{x, p\rightarrow p} } + \epsilon \\
(3) \; X_{p} = G_{i}\cdot {\beta_{x, i\rightarrow p} } + \epsilon \\
(4) \; Y_{i} = G_{i}\cdot {\beta_{y, i\rightarrow i} } + \epsilon \\
(5) \; Y_{p} = G_{i}\cdot {\beta_{y, i\rightarrow p} } + \epsilon \\
(6) \; Y_{i} = \alpha_{x_{i} \rightarrow y_{i}}\cdot X_{i} +\epsilon \\
(7) \; Y_{p} = \alpha_{x_{p} \rightarrow y_{p}}\cdot X_{p} +\epsilon \\
(8) \; X_{p} = \alpha_{x_{i} \rightarrow x_{p}}\cdot X_{i{}} +\epsilon \\
(9) \; Y_{p} = \alpha_{y_{i} \rightarrow y_{p}}\cdot Y_{i{}} +\epsilon \\
(10) \; Y_{p} = \alpha_{x_{i} \rightarrow y_{p}}\cdot X_{i} +\epsilon
\end{align}
$$

- Equation (1), (2), and (4) are simply same-person GWAS result (obtained from Neale).
- Equation (3) and (5) are household GWAS results (calculated in-house).
- Equation (6) and (7) (${\alpha_{x_{i} \rightarrow y_{i}}}$ and $\alpha_{x_{p} \rightarrow y_{p}}$), in the causal effect of a trait ${X}$ on ${Y}$ from a standard MR single-person MR, sex specific.
- Equation (8) and (9), represent assortative mating effects for traits ${X}$ and ${Y}$, respectively ($\alpha_{x_{i} \rightarrow x_{p}}$ and $\alpha_{y_{i} \rightarrow y_{p}}$).
- Equation (10) $\alpha_{x_{i} \rightarrow y_{p}}$ is the result of phenotype ${X}$ in the index, having a causal effect on a phenotype ${Y}$ in the partner. 
    - This effect is either due to:
        (1) The product of $\alpha_{x_{i} \rightarrow x_{p}}$ with $\alpha_{x_{p} \rightarrow y_{p}}$, or 
        (2) The product of $\alpha_{x_{i} \rightarrow y_{i}}$ with $\alpha_{y_{i} \rightarrow y_{p}}$, or
        (3) An additional 3rd (or 4th, etc.) variable which amplifies the effect from ${X_{i}}$ to ${Y_{p}}$ that doesn't simply act through a combination of assortative mating and a causal effect of $X$ on $Y$.
    
